// gifted-charts-core/dist/LineChart/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useEffect, useMemo, useState } from 'react';
import { AxesAndRulesDefaults, LineDefaults, SEGMENT_END, SEGMENT_START, chartTypes, defaultArrowConfig, defaultPointerConfig } from '../utils/constants';
import { colorsToLowerCase, computeMaxAndMinItems, getAllArrowProperties, getArrowPoints, getAxesAndRulesProps, getCurvePathWithSegments, getExtendedContainerHeightWithPadding, getLineSegmentsDueToNoExtrapolation, getLineSegmentsForMissingValues, getMaxValue, getNoOfSections, getPathWithHighlight, getSanitisedData, getSecondaryDataWithOffsetIncluded, getSegmentString, indexOfFirstNonZeroDigit, svgPath } from '../utils';
import { EdgePosition } from '../utils/types';
export var useLineChart = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66, _67, _68, _69, _70, _71, _72, _73, _74, _75, _76, _77, _78, _79, _80, _81, _82, _83, _84, _85, _86, _87, _88, _89, _90, _91, _92, _93, _94, _95, _96, _97, _98, _99, _100, _101, _102, _103, _104, _105, _106, _107, _108, _109, _110, _111, _112, _113, _114, _115, _116, _117, _118, _119, _120, _121, _122, _123, _124, _125, _126, _127, _128, _129, _130, _131, _132, _133, _134, _135, _136, _137, _138, _139, _140, _141, _142, _143, _144, _145, _146, _147, _148, _149, _150, _151, _152, _153, _154, _155, _156, _157, _158, _159, _160, _161, _162, _163, _164, _165, _166, _167, _168, _169, _170, _171, _172, _173, _174, _175, _176, _177, _178, _179, _180, _181, _182, _183, _184, _185, _186, _187, _188, _189, _190, _191, _192, _193, _194, _195, _196, _197, _198, _199, _200, _201, _202, _203, _204, _205, _206, _207, _208, _209, _210, _211, _212, _213, _214, _215, _216, _217, _218, _219, _220, _221, _222, _223, _224, _225, _226, _227, _228, _229, _230, _231, _232, _233, _234, _235, _236, _237, _238, _239, _240, _241, _242, _243, _244, _245, _246, _247, _248, _249, _250, _251, _252, _253, _254, _255, _256, _257, _258, _259, _260, _261, _262, _263, _264, _265, _266, _267, _268, _269, _270, _271, _272, _273, _274, _275, _276, _277, _278, _279, _280, _281, _282, _283, _284, _285, _286, _287, _288, _289, _290, _291, _292, _293, _294, _295, _296, _297, _298, _299, _300, _301, _302, _303, _304, _305, _306, _307, _308, _309, _310, _311, _312, _313, _314, _315, _316, _317, _318, _319, _320, _321, _322, _323, _324, _325, _326, _327, _328, _329, _330, _331, _332, _333, _334, _335, _336, _337, _338, _339;
    var showDataPointsForMissingValues = props.showDataPointsForMissingValues, _340 = props.interpolateMissingValues, interpolateMissingValues = _340 === void 0 ? true : _340, _341 = props.extrapolateMissingValues, extrapolateMissingValues = _341 === void 0 ? true : _341, yAxisOffset = props.yAxisOffset, parentWidth = props.parentWidth, renderTooltip = props.renderTooltip, _342 = props.renderTooltip1, renderTooltip1 = _342 === void 0 ? props.renderTooltip : _342, _343 = props.renderTooltip2, renderTooltip2 = _343 === void 0 ? props.renderTooltip : _343, _344 = props.renderTooltip3, renderTooltip3 = _344 === void 0 ? props.renderTooltip : _344, _345 = props.renderTooltip4, renderTooltip4 = _345 === void 0 ? props.renderTooltip : _345, _346 = props.renderTooltip5, renderTooltip5 = _346 === void 0 ? props.renderTooltip : _346, _347 = props.renderTooltipSecondary, renderTooltipSecondary = _347 === void 0 ? props.renderTooltip : _347;
    var lastLineNumber = 1;
    if (props.secondaryData) {
        lastLineNumber = 6667; // lastLineNumber is 6667 for a secondary line, so the index or key of the secondary line is 6666
    }
    if (props.data2)
        lastLineNumber = 2;
    if (props.data3)
        lastLineNumber = 3;
    if (props.data4)
        lastLineNumber = 4;
    if (props.data5)
        lastLineNumber = 5;
    if (((_b = (_a = props.dataSet) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > lastLineNumber)
        lastLineNumber = (_d = (_c = props.dataSet) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0;
    var containsNegativeValue = ((_e = props.mostNegativeValue) !== null && _e !== void 0 ? _e : 0) < 0 ||
        ((_j = ((_h = (_g = (_f = props.dataSet) === null || _f === void 0 ? void 0 : _f[0]) === null || _g === void 0 ? void 0 : _g.data) !== null && _h !== void 0 ? _h : props.data)) === null || _j === void 0 ? void 0 : _j.some(function (item) { var _a; return ((_a = item.value) !== null && _a !== void 0 ? _a : 0) < 0; }));
    var onlyPositive = (_k = props.onlyPositive) !== null && _k !== void 0 ? _k : (!extrapolateMissingValues ? (containsNegativeValue ? false : true) : false);
    // the default value of onlyPositive is usually supposed to be false
    // but we are setting it to true if extrapolateMissingValues is false and there are no negative values
    // because in absence of extrapolation we don't want the chart to unnecessarily show the 4th quadrant
    var curvature = (_l = props.curvature) !== null && _l !== void 0 ? _l : LineDefaults.curvature;
    var curveType = (_m = props.curveType) !== null && _m !== void 0 ? _m : LineDefaults.curveType;
    var _348 = __read(useState(0), 2), scrollX = _348[0], setScrollX = _348[1];
    var _349 = __read(useState(''), 2), arrow1Points = _349[0], setArrow1Points = _349[1];
    var _350 = __read(useState(''), 2), arrow2Points = _350[0], setArrow2Points = _350[1];
    var _351 = __read(useState(''), 2), arrow3Points = _351[0], setArrow3Points = _351[1];
    var _352 = __read(useState(''), 2), arrow4Points = _352[0], setArrow4Points = _352[1];
    var _353 = __read(useState(''), 2), arrow5Points = _353[0], setArrow5Points = _353[1];
    var _354 = __read(useState(''), 2), secondaryArrowPoints = _354[0], setSecondaryArrowPoints = _354[1];
    var _355 = __read(useState(-1), 2), pointerIndex = _355[0], setPointerIndex = _355[1];
    var _356 = __read(useState(0), 2), pointerX = _356[0], setPointerX = _356[1];
    var _357 = __read(useState(0), 2), pointerY = _357[0], setPointerY = _357[1];
    var _358 = __read(useState(), 2), pointerItem = _358[0], setPointerItem = _358[1];
    var _359 = __read(useState(0), 2), pointerY2 = _359[0], setPointerY2 = _359[1];
    var _360 = __read(useState(), 2), pointerItem2 = _360[0], setPointerItem2 = _360[1];
    var _361 = __read(useState(0), 2), pointerY3 = _361[0], setPointerY3 = _361[1];
    var _362 = __read(useState(), 2), pointerItem3 = _362[0], setPointerItem3 = _362[1];
    var _363 = __read(useState(0), 2), pointerY4 = _363[0], setPointerY4 = _363[1];
    var _364 = __read(useState(), 2), pointerItem4 = _364[0], setPointerItem4 = _364[1];
    var _365 = __read(useState(0), 2), pointerY5 = _365[0], setPointerY5 = _365[1];
    var _366 = __read(useState([]), 2), pointerYsForDataSet = _366[0], setPointerYsForDataSet = _366[1];
    var _367 = __read(useState(), 2), pointerItem5 = _367[0], setPointerItem5 = _367[1];
    var _368 = __read(useState(0), 2), secondaryPointerY = _368[0], setSecondaryPointerY = _368[1];
    var _369 = __read(useState(), 2), secondaryPointerItem = _369[0], setSecondaryPointerItem = _369[1];
    var _370 = __read(useState([]), 2), pointerItemsForSet = _370[0], setPointerItemsForSet = _370[1];
    var _371 = __read(useState([]), 2), secondaryPointerItemsForSet = _371[0], setSecondaryPointerItemsForSet = _371[1];
    var _372 = __read(useState(0), 2), responderStartTime = _372[0], setResponderStartTime = _372[1];
    var _373 = __read(useState(false), 2), responderActive = _373[0], setResponderActive = _373[1];
    var _374 = __read(useState(''), 2), points = _374[0], setPoints = _374[1];
    var _375 = __read(useState(''), 2), points2 = _375[0], setPoints2 = _375[1];
    var _376 = __read(useState(''), 2), points3 = _376[0], setPoints3 = _376[1];
    var _377 = __read(useState(''), 2), points4 = _377[0], setPoints4 = _377[1];
    var _378 = __read(useState(''), 2), points5 = _378[0], setPoints5 = _378[1];
    var _379 = __read(useState(''), 2), secondaryPoints = _379[0], setSecondaryPoints = _379[1];
    var _380 = __read(useState(''), 2), fillPoints = _380[0], setFillPoints = _380[1];
    var _381 = __read(useState(''), 2), fillPoints2 = _381[0], setFillPoints2 = _381[1];
    var _382 = __read(useState(''), 2), fillPoints3 = _382[0], setFillPoints3 = _382[1];
    var _383 = __read(useState(''), 2), fillPoints4 = _383[0], setFillPoints4 = _383[1];
    var _384 = __read(useState(''), 2), fillPoints5 = _384[0], setFillPoints5 = _384[1];
    var _385 = __read(useState(''), 2), secondaryFillPoints = _385[0], setSecondaryFillPoints = _385[1];
    var _386 = __read(useState([]), 2), pointsFromSet = _386[0], setPointsFromSet = _386[1];
    var _387 = __read(useState([]), 2), fillPointsFromSet = _387[0], setFillPointsFromSet = _387[1];
    var _388 = __read(useState([]), 2), arrowPointsFromSet = _388[0], setArrowPointsFromSet = _388[1];
    var _389 = __read(useState((_o = props.focusedDataPointIndex) !== null && _o !== void 0 ? _o : -1), 2), selectedIndex = _389[0], setSelectedIndex = _389[1];
    var allowFontScaling = (_p = props.allowFontScaling) !== null && _p !== void 0 ? _p : AxesAndRulesDefaults.allowFontScaling;
    useEffect(function () {
        var _a;
        setSelectedIndex((_a = props.focusedDataPointIndex) !== null && _a !== void 0 ? _a : -1);
    }, [props.focusedDataPointIndex]);
    var noOfSections = getNoOfSections(props.noOfSections, props.maxValue, props.stepValue);
    var containerHeight = (_q = props.height) !== null && _q !== void 0 ? _q : (((_r = props.stepHeight) !== null && _r !== void 0 ? _r : 0) * noOfSections ||
        AxesAndRulesDefaults.containerHeight);
    var dataSanitisationProps = {
        showDataPointsForMissingValues: showDataPointsForMissingValues,
        interpolateMissingValues: interpolateMissingValues,
        onlyPositive: onlyPositive,
        yAxisOffset: yAxisOffset
    };
    var data = useMemo(function () { return getSanitisedData(props.data, dataSanitisationProps); }, [yAxisOffset, props.data]);
    var data2 = useMemo(function () { return getSanitisedData(props.data2, dataSanitisationProps); }, [yAxisOffset, props.data2]);
    var data3 = useMemo(function () { return getSanitisedData(props.data3, dataSanitisationProps); }, [yAxisOffset, props.data3]);
    var data4 = useMemo(function () { return getSanitisedData(props.data4, dataSanitisationProps); }, [yAxisOffset, props.data4]);
    var data5 = useMemo(function () { return getSanitisedData(props.data5, dataSanitisationProps); }, [yAxisOffset, props.data5]);
    var secondaryData = (_s = getSecondaryDataWithOffsetIncluded(props.secondaryData, props.secondaryYAxis, showDataPointsForMissingValues, interpolateMissingValues, onlyPositive)) !== null && _s !== void 0 ? _s : [];
    var dataSet = undefined;
    if ((_t = props.dataSet) === null || _t === void 0 ? void 0 : _t.length) {
        dataSet = useMemo(function () {
            var _a;
            return (_a = props.dataSet) === null || _a === void 0 ? void 0 : _a.map(function (dataSetItem) {
                var sanitisedData = getSanitisedData(dataSetItem.data, dataSanitisationProps);
                return __assign(__assign({}, dataSetItem), { data: sanitisedData });
            });
        }, [yAxisOffset, props.dataSet]);
    }
    var data0 = (_u = dataSet === null || dataSet === void 0 ? void 0 : dataSet[0]) === null || _u === void 0 ? void 0 : _u.data;
    var scrollToEnd = (_v = props.scrollToEnd) !== null && _v !== void 0 ? _v : LineDefaults.scrollToEnd;
    var scrollAnimation = (_w = props.scrollAnimation) !== null && _w !== void 0 ? _w : LineDefaults.scrollAnimation;
    var scrollEventThrottle = (_x = props.scrollEventThrottle) !== null && _x !== void 0 ? _x : LineDefaults.scrollEventThrottle;
    var labelsExtraHeight = (_y = props.labelsExtraHeight) !== null && _y !== void 0 ? _y : 0;
    var animationDuration = (_z = props.animationDuration) !== null && _z !== void 0 ? _z : LineDefaults.animationDuration;
    var onDataChangeAnimationDuration = (_0 = props.onDataChangeAnimationDuration) !== null && _0 !== void 0 ? _0 : 400;
    var animateTogether = (_1 = props.animateTogether) !== null && _1 !== void 0 ? _1 : LineDefaults.animateTogether;
    var renderDataPointsAfterAnimationEnds = (_2 = props.renderDataPointsAfterAnimationEnds) !== null && _2 !== void 0 ? _2 : LineDefaults.renderDataPointsAfterAnimationEnds;
    var animateOnDataChange = (_3 = props.animateOnDataChange) !== null && _3 !== void 0 ? _3 : false;
    var startIndex1 = (_5 = (_4 = props.startIndex1) !== null && _4 !== void 0 ? _4 : props.startIndex) !== null && _5 !== void 0 ? _5 : 0;
    var endIndex1;
    if (props.endIndex1 === undefined || props.endIndex1 === null) {
        if (props.endIndex === undefined || props.endIndex === null) {
            endIndex1 = data.length - 1;
        }
        else {
            endIndex1 = props.endIndex;
        }
    }
    else {
        endIndex1 = props.endIndex1;
    }
    var startIndex2 = (_6 = props.startIndex2) !== null && _6 !== void 0 ? _6 : 0;
    var endIndex2 = (_7 = props.endIndex2) !== null && _7 !== void 0 ? _7 : data2.length - 1;
    var startIndex3 = (_8 = props.startIndex3) !== null && _8 !== void 0 ? _8 : 0;
    var endIndex3 = (_9 = props.endIndex3) !== null && _9 !== void 0 ? _9 : data3.length - 1;
    var startIndex4 = (_10 = props.startIndex4) !== null && _10 !== void 0 ? _10 : 0;
    var endIndex4 = (_11 = props.endIndex4) !== null && _11 !== void 0 ? _11 : data4.length - 1;
    var startIndex5 = (_12 = props.startIndex5) !== null && _12 !== void 0 ? _12 : 0;
    var endIndex5 = (_13 = props.endIndex5) !== null && _13 !== void 0 ? _13 : data5.length - 1;
    var lengthOfLongestDataArray = Math.max(data.length, data2.length, data3.length, data4.length, data5.length);
    var lineSegments = !interpolateMissingValues
        ? getLineSegmentsForMissingValues(props.data)
        : !extrapolateMissingValues
            ? getLineSegmentsDueToNoExtrapolation(props.data)
            : colorsToLowerCase(props.lineSegments);
    var lineSegments2 = !interpolateMissingValues
        ? getLineSegmentsForMissingValues(props.data2)
        : !extrapolateMissingValues
            ? getLineSegmentsDueToNoExtrapolation(props.data2)
            : colorsToLowerCase(props.lineSegments2);
    var lineSegments3 = !interpolateMissingValues
        ? getLineSegmentsForMissingValues(props.data3)
        : !extrapolateMissingValues
            ? getLineSegmentsDueToNoExtrapolation(props.data3)
            : colorsToLowerCase(props.lineSegments3);
    var lineSegments4 = !interpolateMissingValues
        ? getLineSegmentsForMissingValues(props.data4)
        : !extrapolateMissingValues
            ? getLineSegmentsDueToNoExtrapolation(props.data4)
            : colorsToLowerCase(props.lineSegments4);
    var lineSegments5 = !interpolateMissingValues
        ? getLineSegmentsForMissingValues(props.data5)
        : !extrapolateMissingValues
            ? getLineSegmentsDueToNoExtrapolation(props.data5)
            : colorsToLowerCase(props.lineSegments5);
    var highlightedRange = props.highlightedRange;
    var adjustToWidth = (_14 = props.adjustToWidth) !== null && _14 !== void 0 ? _14 : false;
    var initialSpacing = (_15 = props.initialSpacing) !== null && _15 !== void 0 ? _15 : LineDefaults.initialSpacing;
    var endSpacing = (_16 = props.endSpacing) !== null && _16 !== void 0 ? _16 : (adjustToWidth ? 0 : LineDefaults.endSpacing);
    var thickness = (_17 = props.thickness) !== null && _17 !== void 0 ? _17 : LineDefaults.thickness;
    var yAxisLabelWidth = (_18 = props.yAxisLabelWidth) !== null && _18 !== void 0 ? _18 : (props.hideYAxisText
        ? AxesAndRulesDefaults.yAxisEmptyLabelWidth
        : AxesAndRulesDefaults.yAxisLabelWidth);
    var spacing = (_19 = props.spacing) !== null && _19 !== void 0 ? _19 : (adjustToWidth
        ? (((_20 = props.width) !== null && _20 !== void 0 ? _20 : parentWidth - yAxisLabelWidth) - initialSpacing) /
            Math.max((data0 !== null && data0 !== void 0 ? data0 : data).length - 1, 1)
        : LineDefaults.spacing);
    var xAxisThickness = (_21 = props.xAxisThickness) !== null && _21 !== void 0 ? _21 : AxesAndRulesDefaults.xAxisThickness;
    var dataPointsHeight1 = (_23 = (_22 = props.dataPointsHeight1) !== null && _22 !== void 0 ? _22 : props.dataPointsHeight) !== null && _23 !== void 0 ? _23 : LineDefaults.dataPointsHeight;
    var dataPointsWidth1 = (_25 = (_24 = props.dataPointsWidth1) !== null && _24 !== void 0 ? _24 : props.dataPointsWidth) !== null && _25 !== void 0 ? _25 : LineDefaults.dataPointsWidth;
    var dataPointsRadius1 = (_27 = (_26 = props.dataPointsRadius1) !== null && _26 !== void 0 ? _26 : props.dataPointsRadius) !== null && _27 !== void 0 ? _27 : LineDefaults.dataPointsRadius;
    var dataPointsColor1 = (_29 = (_28 = props.dataPointsColor1) !== null && _28 !== void 0 ? _28 : props.dataPointsColor) !== null && _29 !== void 0 ? _29 : LineDefaults.dataPointsColor;
    var dataPointsShape1 = (_31 = (_30 = props.dataPointsShape1) !== null && _30 !== void 0 ? _30 : props.dataPointsShape) !== null && _31 !== void 0 ? _31 : LineDefaults.dataPointsShape;
    var dataPointsHeight2 = (_33 = (_32 = props.dataPointsHeight2) !== null && _32 !== void 0 ? _32 : props.dataPointsHeight) !== null && _33 !== void 0 ? _33 : LineDefaults.dataPointsHeight;
    var dataPointsWidth2 = (_35 = (_34 = props.dataPointsWidth2) !== null && _34 !== void 0 ? _34 : props.dataPointsWidth) !== null && _35 !== void 0 ? _35 : LineDefaults.dataPointsWidth;
    var dataPointsRadius2 = (_37 = (_36 = props.dataPointsRadius2) !== null && _36 !== void 0 ? _36 : props.dataPointsRadius) !== null && _37 !== void 0 ? _37 : LineDefaults.dataPointsRadius;
    var dataPointsColor2 = (_39 = (_38 = props.dataPointsColor2) !== null && _38 !== void 0 ? _38 : props.dataPointsColor) !== null && _39 !== void 0 ? _39 : LineDefaults.dataPointsColor2;
    var dataPointsShape2 = (_41 = (_40 = props.dataPointsShape2) !== null && _40 !== void 0 ? _40 : props.dataPointsShape) !== null && _41 !== void 0 ? _41 : LineDefaults.dataPointsShape;
    var dataPointsHeight3 = (_43 = (_42 = props.dataPointsHeight3) !== null && _42 !== void 0 ? _42 : props.dataPointsHeight) !== null && _43 !== void 0 ? _43 : LineDefaults.dataPointsHeight;
    var dataPointsWidth3 = (_45 = (_44 = props.dataPointsWidth3) !== null && _44 !== void 0 ? _44 : props.dataPointsWidth) !== null && _45 !== void 0 ? _45 : LineDefaults.dataPointsWidth;
    var dataPointsRadius3 = (_47 = (_46 = props.dataPointsRadius3) !== null && _46 !== void 0 ? _46 : props.dataPointsRadius) !== null && _47 !== void 0 ? _47 : LineDefaults.dataPointsRadius;
    var dataPointsColor3 = (_49 = (_48 = props.dataPointsColor3) !== null && _48 !== void 0 ? _48 : props.dataPointsColor) !== null && _49 !== void 0 ? _49 : LineDefaults.dataPointsColor3;
    var dataPointsShape3 = (_51 = (_50 = props.dataPointsShape3) !== null && _50 !== void 0 ? _50 : props.dataPointsShape) !== null && _51 !== void 0 ? _51 : LineDefaults.dataPointsShape;
    var dataPointsHeight4 = (_53 = (_52 = props.dataPointsHeight4) !== null && _52 !== void 0 ? _52 : props.dataPointsHeight) !== null && _53 !== void 0 ? _53 : LineDefaults.dataPointsHeight;
    var dataPointsWidth4 = (_55 = (_54 = props.dataPointsWidth4) !== null && _54 !== void 0 ? _54 : props.dataPointsWidth) !== null && _55 !== void 0 ? _55 : LineDefaults.dataPointsWidth;
    var dataPointsRadius4 = (_57 = (_56 = props.dataPointsRadius4) !== null && _56 !== void 0 ? _56 : props.dataPointsRadius) !== null && _57 !== void 0 ? _57 : LineDefaults.dataPointsRadius;
    var dataPointsColor4 = (_59 = (_58 = props.dataPointsColor4) !== null && _58 !== void 0 ? _58 : props.dataPointsColor) !== null && _59 !== void 0 ? _59 : LineDefaults.dataPointsColor;
    var dataPointsShape4 = (_61 = (_60 = props.dataPointsShape4) !== null && _60 !== void 0 ? _60 : props.dataPointsShape) !== null && _61 !== void 0 ? _61 : LineDefaults.dataPointsShape;
    var dataPointsHeight5 = (_63 = (_62 = props.dataPointsHeight5) !== null && _62 !== void 0 ? _62 : props.dataPointsHeight) !== null && _63 !== void 0 ? _63 : LineDefaults.dataPointsHeight;
    var dataPointsWidth5 = (_65 = (_64 = props.dataPointsWidth5) !== null && _64 !== void 0 ? _64 : props.dataPointsWidth) !== null && _65 !== void 0 ? _65 : LineDefaults.dataPointsWidth;
    var dataPointsRadius5 = (_67 = (_66 = props.dataPointsRadius5) !== null && _66 !== void 0 ? _66 : props.dataPointsRadius) !== null && _67 !== void 0 ? _67 : LineDefaults.dataPointsRadius;
    var dataPointsColor5 = (_69 = (_68 = props.dataPointsColor5) !== null && _68 !== void 0 ? _68 : props.dataPointsColor) !== null && _69 !== void 0 ? _69 : LineDefaults.dataPointsColor;
    var dataPointsShape5 = (_71 = (_70 = props.dataPointsShape5) !== null && _70 !== void 0 ? _70 : props.dataPointsShape) !== null && _71 !== void 0 ? _71 : LineDefaults.dataPointsShape;
    var areaChart = props.areaChart;
    var areaChart1 = props.areaChart1;
    var areaChart2 = props.areaChart2;
    var areaChart3 = props.areaChart3;
    var areaChart4 = props.areaChart4;
    var areaChart5 = props.areaChart5;
    var atLeastOneAreaChart = (dataSet === null || dataSet === void 0 ? void 0 : dataSet.some(function (set) { return set.areaChart; })) ||
        areaChart ||
        areaChart1 ||
        areaChart2 ||
        areaChart3 ||
        areaChart4 ||
        areaChart5;
    var getIsNthAreaChart = function (n) {
        if (areaChart)
            return true;
        if (!(dataSet === null || dataSet === void 0 ? void 0 : dataSet.length)) {
            switch (n) {
                case 0:
                    return areaChart1 !== null && areaChart1 !== void 0 ? areaChart1 : false;
                case 1:
                    return areaChart2 !== null && areaChart2 !== void 0 ? areaChart2 : false;
                case 2:
                    return areaChart3 !== null && areaChart3 !== void 0 ? areaChart3 : false;
                case 3:
                    return areaChart4 !== null && areaChart4 !== void 0 ? areaChart4 : false;
                case 4:
                    return areaChart5 !== null && areaChart5 !== void 0 ? areaChart5 : false;
            }
        }
        return false;
    };
    var stepChart = (_72 = props.stepChart) !== null && _72 !== void 0 ? _72 : false;
    var stepChart1 = (_73 = props.stepChart1) !== null && _73 !== void 0 ? _73 : false;
    var stepChart2 = (_74 = props.stepChart2) !== null && _74 !== void 0 ? _74 : false;
    var stepChart3 = (_75 = props.stepChart3) !== null && _75 !== void 0 ? _75 : false;
    var stepChart4 = (_76 = props.stepChart4) !== null && _76 !== void 0 ? _76 : false;
    var stepChart5 = (_77 = props.stepChart5) !== null && _77 !== void 0 ? _77 : false;
    var edgePosition = (_78 = props.edgePosition) !== null && _78 !== void 0 ? _78 : LineDefaults.edgePosition;
    var textFontSize1 = (_80 = (_79 = props.textFontSize1) !== null && _79 !== void 0 ? _79 : props.textFontSize) !== null && _80 !== void 0 ? _80 : LineDefaults.textFontSize;
    var textFontSize2 = (_82 = (_81 = props.textFontSize2) !== null && _81 !== void 0 ? _81 : props.textFontSize) !== null && _82 !== void 0 ? _82 : LineDefaults.textFontSize;
    var textFontSize3 = (_84 = (_83 = props.textFontSize3) !== null && _83 !== void 0 ? _83 : props.textFontSize) !== null && _84 !== void 0 ? _84 : LineDefaults.textFontSize;
    var textFontSize4 = (_86 = (_85 = props.textFontSize4) !== null && _85 !== void 0 ? _85 : props.textFontSize) !== null && _86 !== void 0 ? _86 : LineDefaults.textFontSize;
    var textFontSize5 = (_88 = (_87 = props.textFontSize5) !== null && _87 !== void 0 ? _87 : props.textFontSize) !== null && _88 !== void 0 ? _88 : LineDefaults.textFontSize;
    var textColor1 = (_90 = (_89 = props.textColor1) !== null && _89 !== void 0 ? _89 : props.textColor) !== null && _90 !== void 0 ? _90 : LineDefaults.textColor;
    var textColor2 = (_92 = (_91 = props.textColor2) !== null && _91 !== void 0 ? _91 : props.textColor) !== null && _92 !== void 0 ? _92 : LineDefaults.textColor;
    var textColor3 = (_94 = (_93 = props.textColor3) !== null && _93 !== void 0 ? _93 : props.textColor) !== null && _94 !== void 0 ? _94 : LineDefaults.textColor;
    var textColor4 = (_96 = (_95 = props.textColor4) !== null && _95 !== void 0 ? _95 : props.textColor) !== null && _96 !== void 0 ? _96 : LineDefaults.textColor;
    var textColor5 = (_98 = (_97 = props.textColor5) !== null && _97 !== void 0 ? _97 : props.textColor) !== null && _98 !== void 0 ? _98 : LineDefaults.textColor;
    var mergedPrimaryDataArrays = [];
    var mergedPrimaryDataArraysOriginals = [];
    var mergedSecondaryDataArrays = [];
    var maxSpacingSum = 0; // max of spacingSum among all the lines
    var cumulativeSpacing1 = [], cumulativeSpacing2 = [], cumulativeSpacing3 = [], cumulativeSpacing4 = [], cumulativeSpacing5 = [], cumulativeSpacingSecondary = [];
    var cumulativeSpacingForSet = [[]]; // Array(dataSet?.length ?? 0).fill([])
    var strips = {};
    if (dataSet === null || dataSet === void 0 ? void 0 : dataSet.length) {
        dataSet.forEach(function (set, key) {
            var _a, _b, _c, _d;
            if (set.isSecondary) {
                mergedSecondaryDataArrays.push.apply(mergedSecondaryDataArrays, __spreadArray([], __read(set.data), false));
            }
            else {
                mergedPrimaryDataArrays.push.apply(mergedPrimaryDataArrays, __spreadArray([], __read(set.data), false));
                mergedPrimaryDataArraysOriginals.push.apply(mergedPrimaryDataArraysOriginals, __spreadArray([], __read(((_c = (_b = (_a = props.dataSet) === null || _a === void 0 ? void 0 : _a[key]) === null || _b === void 0 ? void 0 : _b.data) !== null && _c !== void 0 ? _c : [])), false));
            }
            var space = (_d = set.spacing) !== null && _d !== void 0 ? _d : spacing;
            var spacingSum = 0;
            var localCumulativeSum = [];
            set.data.forEach(function (item, index) {
                var _a, _b;
                spacingSum += (_a = item.spacing) !== null && _a !== void 0 ? _a : space;
                localCumulativeSum.push(spacingSum);
                if (item.showStrip) {
                    strips[key] = (_b = strips[key]) !== null && _b !== void 0 ? _b : {};
                    strips[key][index] = { item: item, index: index, key: key };
                }
            });
            cumulativeSpacingForSet[key] = localCumulativeSum;
            if (maxSpacingSum < spacingSum) {
                maxSpacingSum = spacingSum;
            }
        });
    }
    else {
        var spacingSum_1 = 0;
        var space_1 = (_99 = props.spacing1) !== null && _99 !== void 0 ? _99 : spacing;
        data.forEach(function (item, index) {
            var _a, _b;
            spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
            cumulativeSpacing1.push(spacingSum_1);
            if (item.showStrip) {
                strips[0] = (_b = strips[0]) !== null && _b !== void 0 ? _b : {};
                strips[0][index] = { item: item, index: index, key: 0 };
            }
        });
        if (maxSpacingSum < spacingSum_1) {
            maxSpacingSum = spacingSum_1;
        }
        if (data2 === null || data2 === void 0 ? void 0 : data2.length) {
            spacingSum_1 = 0;
            space_1 = (_100 = props.spacing2) !== null && _100 !== void 0 ? _100 : spacing;
            data2.forEach(function (item, index) {
                var _a, _b;
                spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
                cumulativeSpacing2.push(spacingSum_1);
                if (item.showStrip) {
                    strips[1] = (_b = strips[1]) !== null && _b !== void 0 ? _b : {};
                    strips[1][index] = { item: item, index: index, key: 1 };
                }
            });
            if (maxSpacingSum < spacingSum_1) {
                maxSpacingSum = spacingSum_1;
            }
        }
        if (data3 === null || data3 === void 0 ? void 0 : data3.length) {
            spacingSum_1 = 0;
            space_1 = (_101 = props.spacing3) !== null && _101 !== void 0 ? _101 : spacing;
            data3.forEach(function (item, index) {
                var _a, _b;
                spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
                cumulativeSpacing3.push(spacingSum_1);
                if (item.showStrip) {
                    strips[2] = (_b = strips[2]) !== null && _b !== void 0 ? _b : {};
                    strips[2][index] = { item: item, index: index, key: 2 };
                }
            });
            if (maxSpacingSum < spacingSum_1) {
                maxSpacingSum = spacingSum_1;
            }
        }
        if (data4 === null || data4 === void 0 ? void 0 : data4.length) {
            spacingSum_1 = 0;
            space_1 = (_102 = props.spacing4) !== null && _102 !== void 0 ? _102 : spacing;
            data4.forEach(function (item, index) {
                var _a, _b;
                spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
                cumulativeSpacing4.push(spacingSum_1);
                if (item.showStrip) {
                    strips[3] = (_b = strips[3]) !== null && _b !== void 0 ? _b : {};
                    strips[3][index] = { item: item, index: index, key: 3 };
                }
            });
            if (maxSpacingSum < spacingSum_1) {
                maxSpacingSum = spacingSum_1;
            }
        }
        if (data5 === null || data5 === void 0 ? void 0 : data5.length) {
            spacingSum_1 = 0;
            space_1 = (_103 = props.spacing5) !== null && _103 !== void 0 ? _103 : spacing;
            data5.forEach(function (item, index) {
                var _a, _b;
                spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
                cumulativeSpacing5.push(spacingSum_1);
                if (item.showStrip) {
                    strips[4] = (_b = strips[4]) !== null && _b !== void 0 ? _b : {};
                    strips[4][index] = { item: item, index: index, key: 4 };
                }
            });
            if (maxSpacingSum < spacingSum_1) {
                maxSpacingSum = spacingSum_1;
            }
        }
        if (secondaryData === null || secondaryData === void 0 ? void 0 : secondaryData.length) {
            spacingSum_1 = 0;
            space_1 = (_105 = (_104 = props.secondaryLineConfig) === null || _104 === void 0 ? void 0 : _104.spacing) !== null && _105 !== void 0 ? _105 : spacing;
            secondaryData.forEach(function (item) {
                var _a;
                spacingSum_1 += (_a = item.spacing) !== null && _a !== void 0 ? _a : space_1;
                cumulativeSpacingSecondary.push(spacingSum_1);
            });
            if (maxSpacingSum < spacingSum_1) {
                maxSpacingSum = spacingSum_1;
            }
        }
        mergedPrimaryDataArrays = data;
        mergedPrimaryDataArraysOriginals = (_106 = props.data) !== null && _106 !== void 0 ? _106 : [];
        mergedSecondaryDataArrays = secondaryData;
    }
    if (!mergedPrimaryDataArrays.length) {
        mergedPrimaryDataArrays = __spreadArray([], __read(mergedSecondaryDataArrays), false);
    }
    var totalWidth = initialSpacing + maxSpacingSum + endSpacing;
    var valuesRange = Math.max.apply(Math, __spreadArray([], __read(mergedPrimaryDataArrays.map(function (i) { return Math.max(i.value, 0); })), false)) - // find the largest +ve number
     Math.min.apply(// find the largest +ve number
    Math, __spreadArray([], __read(mergedPrimaryDataArrays.map(function (i) { return Math.max(i.value, 0); })), false)); // find the smallest +ve number
    var showFractionalValues = (_107 = props.showFractionalValues) !== null && _107 !== void 0 ? _107 : (isFinite(valuesRange) && valuesRange <= 1);
    var roundToDigits = (_108 = props.roundToDigits) !== null && _108 !== void 0 ? _108 : (showFractionalValues ? indexOfFirstNonZeroDigit(valuesRange) + 1 : 0);
    var _390 = computeMaxAndMinItems(mergedPrimaryDataArrays, extrapolateMissingValues, roundToDigits, showFractionalValues, mergedPrimaryDataArraysOriginals), maxItem = _390.maxItem, minItem = _390.minItem;
    var maxValue = getMaxValue(props.maxValue, props.stepValue, noOfSections, maxItem) || 10;
    var mostNegativeValue = (_109 = props.mostNegativeValue) !== null && _109 !== void 0 ? _109 : minItem;
    var overflowTop = ((_110 = props.overflowTop) !== null && _110 !== void 0 ? _110 : props.secondaryXAxis)
        ? AxesAndRulesDefaults.overflowTopWithSecondaryXAxis
        : AxesAndRulesDefaults.overflowTop;
    var extendedContainerHeight = getExtendedContainerHeightWithPadding(containerHeight, overflowTop);
    var secondaryValuesRange = Math.max.apply(Math, __spreadArray([], __read(mergedSecondaryDataArrays.map(function (i) { return Math.max(i.value, 0); })), false)) - // find the largest +ve number
     Math.min.apply(// find the largest +ve number
    Math, __spreadArray([], __read(mergedSecondaryDataArrays.map(function (i) { return Math.max(i.value, 0); })), false)); // find the smallest +ve number
    var showSecondaryFractionalValues = (_112 = (_111 = props.secondaryYAxis) === null || _111 === void 0 ? void 0 : _111.showFractionalValues) !== null && _112 !== void 0 ? _112 : secondaryValuesRange <= 1;
    var secondaryRoundToDigits = (_114 = (_113 = props.secondaryYAxis) === null || _113 === void 0 ? void 0 : _113.roundToDigits) !== null && _114 !== void 0 ? _114 : (showSecondaryFractionalValues
        ? indexOfFirstNonZeroDigit(secondaryValuesRange) + 1
        : 0);
    var _391 = computeMaxAndMinItems(mergedSecondaryDataArrays, extrapolateMissingValues, secondaryRoundToDigits, showSecondaryFractionalValues), secondaryMaxItem = _391.maxItem, secondaryMinItem = _391.minItem;
    var secondaryMaxValue = (_116 = (_115 = props.secondaryYAxis) === null || _115 === void 0 ? void 0 : _115.maxValue) !== null && _116 !== void 0 ? _116 : (secondaryMaxItem || maxValue);
    var getSecondaryY = function (value) {
        return extendedContainerHeight - (value * containerHeight) / secondaryMaxValue;
    };
    var heightUptoXaxis = extendedContainerHeight - xAxisThickness;
    var showValuesAsDataPointsText = (_117 = props.showValuesAsDataPointsText) !== null && _117 !== void 0 ? _117 : LineDefaults.showValuesAsDataPointsText;
    var thickness1 = (_119 = (_118 = props.thickness1) !== null && _118 !== void 0 ? _118 : props.thickness) !== null && _119 !== void 0 ? _119 : LineDefaults.thickness;
    var thickness2 = (_121 = (_120 = props.thickness2) !== null && _120 !== void 0 ? _120 : props.thickness) !== null && _121 !== void 0 ? _121 : LineDefaults.thickness;
    var thickness3 = (_123 = (_122 = props.thickness3) !== null && _122 !== void 0 ? _122 : props.thickness) !== null && _123 !== void 0 ? _123 : LineDefaults.thickness;
    var thickness4 = (_125 = (_124 = props.thickness4) !== null && _124 !== void 0 ? _124 : props.thickness) !== null && _125 !== void 0 ? _125 : LineDefaults.thickness;
    var thickness5 = (_127 = (_126 = props.thickness5) !== null && _126 !== void 0 ? _126 : props.thickness) !== null && _127 !== void 0 ? _127 : LineDefaults.thickness;
    var zIndex1 = (_128 = props.zIndex1) !== null && _128 !== void 0 ? _128 : 0;
    var zIndex2 = (_129 = props.zIndex2) !== null && _129 !== void 0 ? _129 : 0;
    var zIndex3 = (_130 = props.zIndex3) !== null && _130 !== void 0 ? _130 : 0;
    var zIndex4 = (_131 = props.zIndex4) !== null && _131 !== void 0 ? _131 : 0;
    var zIndex5 = (_132 = props.zIndex5) !== null && _132 !== void 0 ? _132 : 0;
    var strokeDashArray1 = (_133 = props.strokeDashArray1) !== null && _133 !== void 0 ? _133 : props.strokeDashArray;
    var strokeDashArray2 = (_134 = props.strokeDashArray2) !== null && _134 !== void 0 ? _134 : props.strokeDashArray;
    var strokeDashArray3 = (_135 = props.strokeDashArray3) !== null && _135 !== void 0 ? _135 : props.strokeDashArray;
    var strokeDashArray4 = (_136 = props.strokeDashArray4) !== null && _136 !== void 0 ? _136 : props.strokeDashArray;
    var strokeDashArray5 = (_137 = props.strokeDashArray5) !== null && _137 !== void 0 ? _137 : props.strokeDashArray;
    var strokeLinecap1 = (_139 = (_138 = props.strokeLinecap1) !== null && _138 !== void 0 ? _138 : props.strokeLinecap) !== null && _139 !== void 0 ? _139 : LineDefaults.strokeLinecap;
    var strokeLinecap2 = (_141 = (_140 = props.strokeLinecap2) !== null && _140 !== void 0 ? _140 : props.strokeLinecap) !== null && _141 !== void 0 ? _141 : LineDefaults.strokeLinecap;
    var strokeLinecap3 = (_143 = (_142 = props.strokeLinecap3) !== null && _142 !== void 0 ? _142 : props.strokeLinecap) !== null && _143 !== void 0 ? _143 : LineDefaults.strokeLinecap;
    var strokeLinecap4 = (_145 = (_144 = props.strokeLinecap4) !== null && _144 !== void 0 ? _144 : props.strokeLinecap) !== null && _145 !== void 0 ? _145 : LineDefaults.strokeLinecap;
    var strokeLinecap5 = (_147 = (_146 = props.strokeLinecap5) !== null && _146 !== void 0 ? _146 : props.strokeLinecap) !== null && _147 !== void 0 ? _147 : LineDefaults.strokeLinecap;
    var rotateLabel = (_148 = props.rotateLabel) !== null && _148 !== void 0 ? _148 : false;
    var isAnimated = (_149 = props.isAnimated) !== null && _149 !== void 0 ? _149 : false;
    var hideDataPoints1 = (_151 = (_150 = props.hideDataPoints) !== null && _150 !== void 0 ? _150 : props.hideDataPoints1) !== null && _151 !== void 0 ? _151 : false;
    var hideDataPoints2 = (_153 = (_152 = props.hideDataPoints) !== null && _152 !== void 0 ? _152 : props.hideDataPoints2) !== null && _153 !== void 0 ? _153 : false;
    var hideDataPoints3 = (_155 = (_154 = props.hideDataPoints) !== null && _154 !== void 0 ? _154 : props.hideDataPoints3) !== null && _155 !== void 0 ? _155 : false;
    var hideDataPoints4 = (_157 = (_156 = props.hideDataPoints) !== null && _156 !== void 0 ? _156 : props.hideDataPoints4) !== null && _157 !== void 0 ? _157 : false;
    var hideDataPoints5 = (_159 = (_158 = props.hideDataPoints) !== null && _158 !== void 0 ? _158 : props.hideDataPoints5) !== null && _159 !== void 0 ? _159 : false;
    var color1 = (_161 = (_160 = props.color1) !== null && _160 !== void 0 ? _160 : props.color) !== null && _161 !== void 0 ? _161 : LineDefaults.color;
    var color2 = (_163 = (_162 = props.color2) !== null && _162 !== void 0 ? _162 : props.color) !== null && _163 !== void 0 ? _163 : LineDefaults.color;
    var color3 = (_165 = (_164 = props.color3) !== null && _164 !== void 0 ? _164 : props.color) !== null && _165 !== void 0 ? _165 : LineDefaults.color;
    var color4 = (_167 = (_166 = props.color4) !== null && _166 !== void 0 ? _166 : props.color) !== null && _167 !== void 0 ? _167 : LineDefaults.color;
    var color5 = (_169 = (_168 = props.color5) !== null && _168 !== void 0 ? _168 : props.color) !== null && _169 !== void 0 ? _169 : LineDefaults.color;
    var startFillColor1 = (_171 = (_170 = props.startFillColor1) !== null && _170 !== void 0 ? _170 : props.startFillColor) !== null && _171 !== void 0 ? _171 : LineDefaults.startFillColor;
    var endFillColor1 = (_173 = (_172 = props.endFillColor1) !== null && _172 !== void 0 ? _172 : props.endFillColor) !== null && _173 !== void 0 ? _173 : LineDefaults.endFillColor;
    var startOpacity = (_174 = props.startOpacity) !== null && _174 !== void 0 ? _174 : LineDefaults.startOpacity;
    var endOpacity = (_175 = props.endOpacity) !== null && _175 !== void 0 ? _175 : LineDefaults.endOpacity;
    var startOpacity1 = (_176 = props.startOpacity1) !== null && _176 !== void 0 ? _176 : startOpacity;
    var endOpacity1 = (_177 = props.endOpacity1) !== null && _177 !== void 0 ? _177 : endOpacity;
    var startFillColor2 = (_179 = (_178 = props.startFillColor2) !== null && _178 !== void 0 ? _178 : props.startFillColor) !== null && _179 !== void 0 ? _179 : LineDefaults.startFillColor;
    var endFillColor2 = (_181 = (_180 = props.endFillColor2) !== null && _180 !== void 0 ? _180 : props.endFillColor) !== null && _181 !== void 0 ? _181 : LineDefaults.endFillColor;
    var startOpacity2 = (_182 = props.startOpacity2) !== null && _182 !== void 0 ? _182 : startOpacity;
    var endOpacity2 = (_183 = props.endOpacity2) !== null && _183 !== void 0 ? _183 : endOpacity;
    var startFillColor3 = (_185 = (_184 = props.startFillColor3) !== null && _184 !== void 0 ? _184 : props.startFillColor) !== null && _185 !== void 0 ? _185 : LineDefaults.startFillColor;
    var endFillColor3 = (_187 = (_186 = props.endFillColor3) !== null && _186 !== void 0 ? _186 : props.endFillColor) !== null && _187 !== void 0 ? _187 : LineDefaults.endFillColor;
    var startOpacity3 = (_188 = props.startOpacity3) !== null && _188 !== void 0 ? _188 : startOpacity;
    var endOpacity3 = (_189 = props.endOpacity3) !== null && _189 !== void 0 ? _189 : endOpacity;
    var startFillColor4 = (_191 = (_190 = props.startFillColor4) !== null && _190 !== void 0 ? _190 : props.startFillColor) !== null && _191 !== void 0 ? _191 : LineDefaults.startFillColor;
    var endFillColor4 = (_193 = (_192 = props.endFillColor4) !== null && _192 !== void 0 ? _192 : props.endFillColor) !== null && _193 !== void 0 ? _193 : LineDefaults.endFillColor;
    var startOpacity4 = (_194 = props.startOpacity4) !== null && _194 !== void 0 ? _194 : startOpacity;
    var endOpacity4 = (_195 = props.endOpacity4) !== null && _195 !== void 0 ? _195 : endOpacity;
    var startFillColor5 = (_197 = (_196 = props.startFillColor5) !== null && _196 !== void 0 ? _196 : props.startFillColor) !== null && _197 !== void 0 ? _197 : LineDefaults.startFillColor;
    var endFillColor5 = (_199 = (_198 = props.endFillColor5) !== null && _198 !== void 0 ? _198 : props.endFillColor) !== null && _199 !== void 0 ? _199 : LineDefaults.endFillColor;
    var startOpacity5 = (_200 = props.startOpacity5) !== null && _200 !== void 0 ? _200 : startOpacity;
    var endOpacity5 = (_201 = props.endOpacity5) !== null && _201 !== void 0 ? _201 : endOpacity;
    defaultArrowConfig.strokeWidth = (_203 = (_202 = dataSet === null || dataSet === void 0 ? void 0 : dataSet[0]) === null || _202 === void 0 ? void 0 : _202.thickness) !== null && _203 !== void 0 ? _203 : thickness1;
    defaultArrowConfig.strokeColor = (_205 = (_204 = dataSet === null || dataSet === void 0 ? void 0 : dataSet[0]) === null || _204 === void 0 ? void 0 : _204.color) !== null && _205 !== void 0 ? _205 : color1;
    var _392 = getAllArrowProperties(props, defaultArrowConfig), arrowLength1 = _392.arrowLength1, arrowWidth1 = _392.arrowWidth1, arrowStrokeWidth1 = _392.arrowStrokeWidth1, arrowStrokeColor1 = _392.arrowStrokeColor1, arrowFillColor1 = _392.arrowFillColor1, showArrowBase1 = _392.showArrowBase1, arrowLength2 = _392.arrowLength2, arrowWidth2 = _392.arrowWidth2, arrowStrokeWidth2 = _392.arrowStrokeWidth2, arrowStrokeColor2 = _392.arrowStrokeColor2, arrowFillColor2 = _392.arrowFillColor2, showArrowBase2 = _392.showArrowBase2, arrowLength3 = _392.arrowLength3, arrowWidth3 = _392.arrowWidth3, arrowStrokeWidth3 = _392.arrowStrokeWidth3, arrowStrokeColor3 = _392.arrowStrokeColor3, arrowFillColor3 = _392.arrowFillColor3, showArrowBase3 = _392.showArrowBase3, arrowLength4 = _392.arrowLength4, arrowWidth4 = _392.arrowWidth4, arrowStrokeWidth4 = _392.arrowStrokeWidth4, arrowStrokeColor4 = _392.arrowStrokeColor4, arrowFillColor4 = _392.arrowFillColor4, showArrowBase4 = _392.showArrowBase4, arrowLength5 = _392.arrowLength5, arrowWidth5 = _392.arrowWidth5, arrowStrokeWidth5 = _392.arrowStrokeWidth5, arrowStrokeColor5 = _392.arrowStrokeColor5, arrowFillColor5 = _392.arrowFillColor5, showArrowBase5 = _392.showArrowBase5, arrowLengthsFromSet = _392.arrowLengthsFromSet, arrowWidthsFromSet = _392.arrowWidthsFromSet, arrowStrokeWidthsFromSet = _392.arrowStrokeWidthsFromSet, arrowStrokeColorsFromSet = _392.arrowStrokeColorsFromSet, arrowFillColorsFromSet = _392.arrowFillColorsFromSet, showArrowBasesFromSet = _392.showArrowBasesFromSet;
    var secondaryLineConfig = {
        zIndex: (_207 = (_206 = props.secondaryLineConfig) === null || _206 === void 0 ? void 0 : _206.zIndex) !== null && _207 !== void 0 ? _207 : zIndex1,
        curved: (_209 = (_208 = props.secondaryLineConfig) === null || _208 === void 0 ? void 0 : _208.curved) !== null && _209 !== void 0 ? _209 : props.curved,
        curvature: (_211 = (_210 = props.secondaryLineConfig) === null || _210 === void 0 ? void 0 : _210.curvature) !== null && _211 !== void 0 ? _211 : curvature,
        curveType: (_213 = (_212 = props.secondaryLineConfig) === null || _212 === void 0 ? void 0 : _212.curveType) !== null && _213 !== void 0 ? _213 : curveType,
        areaChart: (_215 = (_214 = props.secondaryLineConfig) === null || _214 === void 0 ? void 0 : _214.areaChart) !== null && _215 !== void 0 ? _215 : areaChart,
        color: (_217 = (_216 = props.secondaryLineConfig) === null || _216 === void 0 ? void 0 : _216.color) !== null && _217 !== void 0 ? _217 : color1,
        thickness: (_219 = (_218 = props.secondaryLineConfig) === null || _218 === void 0 ? void 0 : _218.thickness) !== null && _219 !== void 0 ? _219 : thickness1,
        zIndex1: (_221 = (_220 = props.secondaryLineConfig) === null || _220 === void 0 ? void 0 : _220.zIndex1) !== null && _221 !== void 0 ? _221 : zIndex1,
        strokeDashArray: (_223 = (_222 = props.secondaryLineConfig) === null || _222 === void 0 ? void 0 : _222.strokeDashArray) !== null && _223 !== void 0 ? _223 : strokeDashArray1,
        strokeLinecap: (_225 = (_224 = props.secondaryLineConfig) === null || _224 === void 0 ? void 0 : _224.strokeLinecap) !== null && _225 !== void 0 ? _225 : strokeLinecap1,
        startIndex: (_227 = (_226 = props.secondaryLineConfig) === null || _226 === void 0 ? void 0 : _226.startIndex) !== null && _227 !== void 0 ? _227 : startIndex1,
        endIndex: (_229 = (_228 = props.secondaryLineConfig) === null || _228 === void 0 ? void 0 : _228.endIndex) !== null && _229 !== void 0 ? _229 : endIndex1,
        hideDataPoints: (_231 = (_230 = props.secondaryLineConfig) === null || _230 === void 0 ? void 0 : _230.hideDataPoints) !== null && _231 !== void 0 ? _231 : hideDataPoints1,
        dataPointsHeight: (_233 = (_232 = props.secondaryLineConfig) === null || _232 === void 0 ? void 0 : _232.dataPointsHeight) !== null && _233 !== void 0 ? _233 : dataPointsHeight1,
        dataPointsWidth: (_235 = (_234 = props.secondaryLineConfig) === null || _234 === void 0 ? void 0 : _234.dataPointsWidth) !== null && _235 !== void 0 ? _235 : dataPointsWidth1,
        dataPointsRadius: (_237 = (_236 = props.secondaryLineConfig) === null || _236 === void 0 ? void 0 : _236.dataPointsRadius) !== null && _237 !== void 0 ? _237 : dataPointsRadius1,
        dataPointsColor: (_239 = (_238 = props.secondaryLineConfig) === null || _238 === void 0 ? void 0 : _238.dataPointsColor) !== null && _239 !== void 0 ? _239 : dataPointsColor1,
        dataPointsShape: (_241 = (_240 = props.secondaryLineConfig) === null || _240 === void 0 ? void 0 : _240.dataPointsShape) !== null && _241 !== void 0 ? _241 : dataPointsShape1,
        showValuesAsDataPointsText: (_243 = (_242 = props.secondaryLineConfig) === null || _242 === void 0 ? void 0 : _242.showValuesAsDataPointsText) !== null && _243 !== void 0 ? _243 : showValuesAsDataPointsText,
        startFillColor: (_245 = (_244 = props.secondaryLineConfig) === null || _244 === void 0 ? void 0 : _244.startFillColor) !== null && _245 !== void 0 ? _245 : startFillColor1,
        endFillColor: (_247 = (_246 = props.secondaryLineConfig) === null || _246 === void 0 ? void 0 : _246.endFillColor) !== null && _247 !== void 0 ? _247 : endFillColor1,
        startOpacity: (_249 = (_248 = props.secondaryLineConfig) === null || _248 === void 0 ? void 0 : _248.startOpacity) !== null && _249 !== void 0 ? _249 : startOpacity1,
        endOpacity: (_251 = (_250 = props.secondaryLineConfig) === null || _250 === void 0 ? void 0 : _250.endOpacity) !== null && _251 !== void 0 ? _251 : endOpacity1,
        textFontSize: (_253 = (_252 = props.secondaryLineConfig) === null || _252 === void 0 ? void 0 : _252.textFontSize) !== null && _253 !== void 0 ? _253 : textFontSize1,
        textColor: (_255 = (_254 = props.secondaryLineConfig) === null || _254 === void 0 ? void 0 : _254.textColor) !== null && _255 !== void 0 ? _255 : textColor1,
        showArrow: (_257 = (_256 = props.secondaryLineConfig) === null || _256 === void 0 ? void 0 : _256.showArrow) !== null && _257 !== void 0 ? _257 : props.showArrows,
        arrowConfig: (_259 = (_258 = props.secondaryLineConfig) === null || _258 === void 0 ? void 0 : _258.arrowConfig) !== null && _259 !== void 0 ? _259 : props.arrowConfig
    };
    var yAxisExtraHeightAtTop = props.trimYAxisAtTop
        ? 0
        : (_260 = props.yAxisExtraHeight) !== null && _260 !== void 0 ? _260 : containerHeight / 20;
    var addLeadingAndTrailingPathForAreaFill = function (initialPath, value, dataLength) {
        return ('M ' +
            initialSpacing +
            ',' +
            heightUptoXaxis +
            ' ' +
            'L ' +
            initialSpacing +
            ',' +
            getY(value) +
            ' ' +
            initialPath +
            ' ' +
            'L ' +
            (initialSpacing + spacing * (dataLength - 1)) +
            ',' +
            heightUptoXaxis +
            ' ' +
            'L ' +
            initialSpacing +
            ',' +
            heightUptoXaxis +
            ' ');
    };
    var getNextPoint = function (data, index, around, before, spacingArray, isSecondary) {
        var isLast = index === data.length - 1;
        return isLast && !(around || before)
            ? ' '
            : ' L' +
                (getX(spacingArray, index) +
                    (around ? (isLast ? 0 : spacing / 2) : before ? 0 : spacing)) +
                ' ' +
                (isSecondary
                    ? getSecondaryY(data[index].value)
                    : getY(data[index].value)) +
                ' ';
    };
    var getStepPath = function (data, i, spacingArray, lineSegment, isSecondary) {
        var around = edgePosition === EdgePosition.AROUND_DATA_POINT;
        var before = edgePosition === EdgePosition.BEFORE_DATA_POINT;
        var isSegment = lineSegment === null || lineSegment === void 0 ? void 0 : lineSegment.find(function (segment) { return segment.startIndex === i; });
        return ('L' +
            (getX(spacingArray, i) -
                (around && i > 0 ? spacing / 2 : before && i > 0 ? spacing : 0)) +
            ' ' +
            (isSecondary ? getSecondaryY(data[i].value) : getY(data[i].value)) + // handle isSecondary for dataSet
            (isSegment
                ? getSegmentString(lineSegment, i, SEGMENT_START, SEGMENT_END, true)
                : '') +
            getNextPoint(data, i, around, before, spacingArray, isSecondary));
    };
    var getSegmentPath = function (data, i, lineSegment, startIndex, endIndex, spacingArray, isSecondary) {
        var path = 'L' +
            getX(spacingArray, i) +
            ' ' +
            (isSecondary ? getSecondaryY(data[i].value) : getY(data[i].value)) +
            ' ' +
            getSegmentString(lineSegment, i, SEGMENT_START, SEGMENT_END);
        if (highlightedRange) {
            path += getPathWithHighlight(data, i, highlightedRange, startIndex, endIndex, spacingArray, getX, getY);
        }
        return path;
    };
    // const [oldPoints, setOldPoints] = useState(points)
    useEffect(function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        if (dataSet) {
            var pointsArray_1 = [];
            var arrowPointsArray_1 = [];
            var fillPointsArray_1 = [];
            dataSet.map(function (set, index) {
                var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
                var setSegments = !interpolateMissingValues
                    ? getLineSegmentsForMissingValues((_a = props.dataSet) === null || _a === void 0 ? void 0 : _a[index].data)
                    : !extrapolateMissingValues
                        ? getLineSegmentsDueToNoExtrapolation((_b = props.dataSet) === null || _b === void 0 ? void 0 : _b[index].data)
                        : set.lineSegments;
                if ((_c = set.curved) !== null && _c !== void 0 ? _c : props.curved) {
                    var pArray = [];
                    for (var i = 0; i < set.data.length; i++) {
                        if (i >= ((_d = set.startIndex) !== null && _d !== void 0 ? _d : 0) &&
                            i <= ((_e = set.endIndex) !== null && _e !== void 0 ? _e : set.data.length - 1)) {
                            pArray.push([
                                getX(cumulativeSpacingForSet[index], i),
                                set.isSecondary
                                    ? getSecondaryY(set.data[i].value)
                                    : getY(set.data[i].value)
                            ]);
                        }
                    }
                    var xx = svgPath(pArray, (_f = set.curveType) !== null && _f !== void 0 ? _f : curveType, (_g = set.curvature) !== null && _g !== void 0 ? _g : curvature);
                    pointsArray_1.push(getCurvePathWithSegments(xx, setSegments, //set.lineSegments,
                    SEGMENT_START, SEGMENT_END, curveType));
                    // For Arrow-
                    if (set.data.length > 1 && ((_h = set.showArrow) !== null && _h !== void 0 ? _h : props.showArrows)) {
                        var arrowTipY = pArray[pArray.length - 1][1];
                        var arrowTipX = pArray[pArray.length - 1][0];
                        var y1 = pArray[pArray.length - 2][1];
                        var x1 = pArray[pArray.length - 2][0];
                        var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLengthsFromSet === null || arrowLengthsFromSet === void 0 ? void 0 : arrowLengthsFromSet[index], arrowWidthsFromSet === null || arrowWidthsFromSet === void 0 ? void 0 : arrowWidthsFromSet[index], showArrowBasesFromSet === null || showArrowBasesFromSet === void 0 ? void 0 : showArrowBasesFromSet[index]);
                        arrowPointsArray_1.push(arrowPoints);
                    }
                    // For Area charts-
                    if (((_j = set.areaChart) !== null && _j !== void 0 ? _j : areaChart) && set.data.length > 0) {
                        xx = addLeadingAndTrailingPathForAreaFill(xx, set.data[0].value, set.data.length);
                        fillPointsArray_1.push(xx);
                    }
                }
                else {
                    var pp = '';
                    for (var i = 0; i < set.data.length; i++) {
                        if (i >= ((_k = set.startIndex) !== null && _k !== void 0 ? _k : 0) &&
                            i <= ((_l = set.endIndex) !== null && _l !== void 0 ? _l : set.data.length - 1)) {
                            if ((_m = set.stepChart) !== null && _m !== void 0 ? _m : stepChart) {
                                pp += getStepPath(set.data, i, cumulativeSpacingForSet[index], setSegments, set.isSecondary);
                            }
                            else {
                                pp += getSegmentPath(set.data, i, setSegments, //set.lineSegments,
                                (_o = set.startIndex) !== null && _o !== void 0 ? _o : 0, (_p = set.endIndex) !== null && _p !== void 0 ? _p : set.data.length - 1, cumulativeSpacingForSet[index], set.isSecondary);
                            }
                        }
                    }
                    pointsArray_1.push(pp.replace('L', 'M'));
                    // For Arrow-
                    if (set.data.length > 1 && ((_q = set.showArrow) !== null && _q !== void 0 ? _q : props.showArrows)) {
                        var ppArray = pp.trim().split(' ');
                        var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                        var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                        var y1 = parseInt(ppArray[ppArray.length - 3]);
                        var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                        var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLengthsFromSet === null || arrowLengthsFromSet === void 0 ? void 0 : arrowLengthsFromSet[index], arrowWidthsFromSet === null || arrowWidthsFromSet === void 0 ? void 0 : arrowWidthsFromSet[index], showArrowBasesFromSet === null || showArrowBasesFromSet === void 0 ? void 0 : showArrowBasesFromSet[index]);
                        arrowPointsArray_1.push(arrowPoints);
                    }
                    // For Area charts-
                    if (((_r = set.areaChart) !== null && _r !== void 0 ? _r : areaChart) && set.data.length > 0) {
                        var ppp = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp += pp;
                        ppp +=
                            'L' +
                                (initialSpacing + spacing * (set.data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        fillPointsArray_1.push(ppp.replace('L', 'M'));
                    }
                }
                return null;
            });
            setPointsFromSet(pointsArray_1);
            setArrowPointsFromSet(arrowPointsArray_1);
            setFillPointsFromSet(fillPointsArray_1);
        }
        else {
            var pp = '';
            var pp2 = '';
            var pp3 = '';
            var pp4 = '';
            var pp5 = '';
            if (!props.curved) {
                for (var i = 0; i < lengthOfLongestDataArray; i++) {
                    if (i >= startIndex1 && i <= endIndex1) {
                        if (stepChart !== null && stepChart !== void 0 ? stepChart : stepChart1) {
                            pp += getStepPath(data, i, cumulativeSpacing1, lineSegments);
                        }
                        else {
                            pp += getSegmentPath(data, i, lineSegments, startIndex1, endIndex1, cumulativeSpacing1);
                        }
                    }
                    if (data2.length > 0 && i >= startIndex2 && i <= endIndex2) {
                        if (stepChart !== null && stepChart !== void 0 ? stepChart : stepChart2) {
                            pp2 += getStepPath(data2, i, cumulativeSpacing2, lineSegments2);
                        }
                        else {
                            pp2 += getSegmentPath(data2, i, lineSegments2, startIndex2, endIndex2, cumulativeSpacing2);
                        }
                    }
                    if (data3.length > 0 && i >= startIndex3 && i <= endIndex3) {
                        if (stepChart !== null && stepChart !== void 0 ? stepChart : stepChart3) {
                            pp3 += getStepPath(data3, i, cumulativeSpacing3, lineSegments3);
                        }
                        else {
                            pp3 += getSegmentPath(data3, i, lineSegments3, startIndex3, endIndex3, cumulativeSpacing3);
                        }
                    }
                    if (data4.length > 0 && i >= startIndex4 && i <= endIndex4) {
                        if (stepChart !== null && stepChart !== void 0 ? stepChart : stepChart4) {
                            pp4 += getStepPath(data4, i, cumulativeSpacing4, lineSegments4);
                        }
                        else {
                            pp4 += getSegmentPath(data4, i, lineSegments4, startIndex4, endIndex4, cumulativeSpacing4);
                        }
                    }
                    if (data5.length > 0 && i >= startIndex5 && i <= endIndex5) {
                        if (stepChart !== null && stepChart !== void 0 ? stepChart : stepChart5) {
                            pp5 += getStepPath(data5, i, cumulativeSpacing5, lineSegments5);
                        }
                        else {
                            pp5 += getSegmentPath(data5, i, lineSegments5, startIndex5, endIndex5, cumulativeSpacing5);
                        }
                    }
                }
                setPoints2(pp2.replace('L', 'M'));
                setPoints3(pp3.replace('L', 'M'));
                setPoints4(pp4.replace('L', 'M'));
                setPoints5(pp5.replace('L', 'M'));
                // if (animateOnDataChange) {
                //   setOldPoints(points)
                // }
                setPoints(pp.replace('L', 'M'));
                if (data.length > 1 && ((_a = props.showArrow1) !== null && _a !== void 0 ? _a : props.showArrows)) {
                    var ppArray = pp.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength1, arrowWidth1, showArrowBase1);
                    setArrow1Points(arrowPoints);
                }
                if (data2.length > 1 && ((_b = props.showArrow2) !== null && _b !== void 0 ? _b : props.showArrows)) {
                    var ppArray = pp2.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength2, arrowWidth2, showArrowBase2);
                    setArrow2Points(arrowPoints);
                }
                if (data3.length > 1 && ((_c = props.showArrow3) !== null && _c !== void 0 ? _c : props.showArrows)) {
                    var ppArray = pp3.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength3, arrowWidth3, showArrowBase3);
                    setArrow3Points(arrowPoints);
                }
                if (data4.length > 1 && ((_d = props.showArrow4) !== null && _d !== void 0 ? _d : props.showArrows)) {
                    var ppArray = pp4.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength4, arrowWidth4, showArrowBase4);
                    setArrow4Points(arrowPoints);
                }
                if (data5.length > 1 && ((_e = props.showArrow5) !== null && _e !== void 0 ? _e : props.showArrows)) {
                    var ppArray = pp5.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength5, arrowWidth5, showArrowBase5);
                    setArrow5Points(arrowPoints);
                }
                /** *************************          For Area Charts          *************************/
                if (atLeastOneAreaChart) {
                    var ppp = '';
                    var ppp2 = '';
                    var ppp3 = '';
                    var ppp4 = '';
                    var ppp5 = '';
                    if ((areaChart1 !== null && areaChart1 !== void 0 ? areaChart1 : areaChart) && data.length > 0) {
                        ppp = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp += pp;
                        ppp +=
                            'L' +
                                (initialSpacing + spacing * (data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        setFillPoints(ppp.replace('L', 'M'));
                    }
                    if ((areaChart2 !== null && areaChart2 !== void 0 ? areaChart2 : areaChart) && data2.length > 0) {
                        ppp2 = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp2 += pp2;
                        ppp2 +=
                            'L' +
                                (initialSpacing + spacing * (data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp2 += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        setFillPoints2(ppp2.replace('L', 'M'));
                    }
                    if ((areaChart3 !== null && areaChart3 !== void 0 ? areaChart3 : areaChart) && data3.length > 0) {
                        ppp3 = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp3 += pp3;
                        ppp3 +=
                            'L' +
                                (initialSpacing + spacing * (data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp3 += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        setFillPoints3(ppp3.replace('L', 'M'));
                    }
                    if ((areaChart4 !== null && areaChart4 !== void 0 ? areaChart4 : areaChart) && data4.length > 0) {
                        ppp4 = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp4 += pp4;
                        ppp4 +=
                            'L' +
                                (initialSpacing + spacing * (data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp4 += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        setFillPoints4(ppp4.replace('L', 'M'));
                    }
                    if ((areaChart5 !== null && areaChart5 !== void 0 ? areaChart5 : areaChart) && data5.length > 0) {
                        ppp5 = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        ppp5 += pp5;
                        ppp5 +=
                            'L' +
                                (initialSpacing + spacing * (data.length - 1)) +
                                ' ' +
                                heightUptoXaxis;
                        ppp5 += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                        setFillPoints5(ppp5.replace('L', 'M'));
                    }
                }
                /*************************************************************************************/
            }
            else {
                var p1Array = [];
                var p2Array = [];
                var p3Array = [];
                var p4Array = [];
                var p5Array = [];
                for (var i = 0; i < lengthOfLongestDataArray; i++) {
                    if (i >= startIndex1 && i <= endIndex1) {
                        p1Array.push([getX(cumulativeSpacing1, i), getY(data[i].value)]);
                    }
                    if (data2.length > 0 && i >= startIndex2 && i <= endIndex2) {
                        p2Array.push([getX(cumulativeSpacing2, i), getY(data2[i].value)]);
                    }
                    if (data3.length > 0 && i >= startIndex3 && i <= endIndex3) {
                        p3Array.push([getX(cumulativeSpacing3, i), getY(data3[i].value)]);
                    }
                    if (data4.length > 0 && i >= startIndex4 && i <= endIndex4) {
                        p4Array.push([getX(cumulativeSpacing4, i), getY(data4[i].value)]);
                    }
                    if (data5.length > 0 && i >= startIndex5 && i <= endIndex5) {
                        p5Array.push([getX(cumulativeSpacing5, i), getY(data5[i].value)]);
                    }
                }
                var xx = svgPath(p1Array, curveType, curvature);
                var xx2 = svgPath(p2Array, curveType, curvature);
                var xx3 = svgPath(p3Array, curveType, curvature);
                var xx4 = svgPath(p4Array, curveType, curvature);
                var xx5 = svgPath(p5Array, curveType, curvature);
                setPoints(getCurvePathWithSegments(xx, lineSegments, SEGMENT_START, SEGMENT_END, curveType));
                setPoints2(getCurvePathWithSegments(xx2, lineSegments2, SEGMENT_START, SEGMENT_END, curveType));
                setPoints3(getCurvePathWithSegments(xx3, lineSegments3, SEGMENT_START, SEGMENT_END, curveType));
                setPoints4(getCurvePathWithSegments(xx4, lineSegments4, SEGMENT_START, SEGMENT_END, curveType));
                setPoints5(getCurvePathWithSegments(xx5, lineSegments5, SEGMENT_START, SEGMENT_END, curveType));
                if (data.length > 1 && ((_f = props.showArrow1) !== null && _f !== void 0 ? _f : props.showArrows)) {
                    var arrowTipY = p1Array[p1Array.length - 1][1];
                    var arrowTipX = p1Array[p1Array.length - 1][0];
                    var y1 = p1Array[p1Array.length - 2][1];
                    var x1 = p1Array[p1Array.length - 2][0];
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength1, arrowWidth1, showArrowBase1);
                    setArrow1Points(arrowPoints);
                }
                if (data2.length > 1 && ((_g = props.showArrow2) !== null && _g !== void 0 ? _g : props.showArrows)) {
                    var arrowTipY = p2Array[p2Array.length - 1][1];
                    var arrowTipX = p2Array[p2Array.length - 1][0];
                    var y1 = p2Array[p2Array.length - 2][1];
                    var x1 = p2Array[p2Array.length - 2][0];
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength2, arrowWidth2, showArrowBase2);
                    setArrow2Points(arrowPoints);
                }
                if (data3.length > 1 && ((_h = props.showArrow3) !== null && _h !== void 0 ? _h : props.showArrows)) {
                    var arrowTipY = p3Array[p3Array.length - 1][1];
                    var arrowTipX = p3Array[p3Array.length - 1][0];
                    var y1 = p3Array[p3Array.length - 2][1];
                    var x1 = p3Array[p3Array.length - 2][0];
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength3, arrowWidth3, showArrowBase3);
                    setArrow2Points(arrowPoints);
                }
                if (data4.length > 1 && ((_j = props.showArrow4) !== null && _j !== void 0 ? _j : props.showArrows)) {
                    var arrowTipY = p4Array[p4Array.length - 1][1];
                    var arrowTipX = p4Array[p4Array.length - 1][0];
                    var y1 = p4Array[p4Array.length - 2][1];
                    var x1 = p4Array[p4Array.length - 2][0];
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength4, arrowWidth4, showArrowBase4);
                    setArrow2Points(arrowPoints);
                }
                if (data5.length > 1 && ((_k = props.showArrow5) !== null && _k !== void 0 ? _k : props.showArrows)) {
                    var arrowTipY = p5Array[p5Array.length - 1][1];
                    var arrowTipX = p5Array[p5Array.length - 1][0];
                    var y1 = p5Array[p5Array.length - 2][1];
                    var x1 = p5Array[p5Array.length - 2][0];
                    var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength5, arrowWidth5, showArrowBase5);
                    setArrow2Points(arrowPoints);
                }
                /** *************************          For Area Charts          *************************/
                if (atLeastOneAreaChart) {
                    if ((areaChart1 !== null && areaChart1 !== void 0 ? areaChart1 : areaChart) && data.length > 0) {
                        xx = addLeadingAndTrailingPathForAreaFill(xx, data[0].value, data.length);
                        setFillPoints(xx);
                    }
                    if ((areaChart2 !== null && areaChart2 !== void 0 ? areaChart2 : areaChart) && data2.length > 0) {
                        xx2 = addLeadingAndTrailingPathForAreaFill(xx2, data2[0].value, data2.length);
                        setFillPoints2(xx2);
                    }
                    if ((areaChart3 !== null && areaChart3 !== void 0 ? areaChart3 : areaChart) && data3.length > 0) {
                        xx3 = addLeadingAndTrailingPathForAreaFill(xx3, data3[0].value, data3.length);
                        setFillPoints3(xx3);
                    }
                    if ((areaChart4 !== null && areaChart4 !== void 0 ? areaChart4 : areaChart) && data4.length > 0) {
                        xx4 = addLeadingAndTrailingPathForAreaFill(xx4, data4[0].value, data4.length);
                        setFillPoints4(xx4);
                    }
                    if ((areaChart5 !== null && areaChart5 !== void 0 ? areaChart5 : areaChart) && data5.length > 0) {
                        xx5 = addLeadingAndTrailingPathForAreaFill(xx5, data5[0].value, data5.length);
                        setFillPoints5(xx5);
                    }
                }
                /*************************************************************************************/
            }
        }
    }, [
        animateOnDataChange,
        areaChart,
        areaChart1,
        areaChart2,
        containerHeight,
        data,
        data2,
        data3,
        data4,
        data5,
        // data0,
        dataPointsWidth1,
        dataPointsWidth2,
        dataPointsWidth3,
        dataPointsWidth4,
        dataPointsWidth5,
        initialSpacing,
        maxValue,
        props.curved,
        spacing,
        xAxisThickness,
        startIndex1,
        endIndex1,
        startIndex2,
        endIndex2,
        startIndex3,
        endIndex3,
        startIndex4,
        endIndex4,
        startIndex5,
        endIndex5,
        arrowLength1,
        arrowWidth1,
        showArrowBase1,
        props.showArrow1,
        props.showArrows,
        props.showArrow2,
        props.showArrow3,
        props.showArrow4,
        props.showArrow5,
        arrowLength2,
        arrowWidth2,
        showArrowBase2,
        arrowLength3,
        arrowWidth3,
        showArrowBase3,
        arrowLength4,
        arrowWidth4,
        showArrowBase4,
        arrowLength5,
        arrowWidth5,
        showArrowBase5
    ]);
    useEffect(function () {
        var _a, _b, _c, _d;
        var pp = '';
        if (!secondaryLineConfig.curved) {
            for (var i = 0; i < secondaryData.length; i++) {
                if (i >= secondaryLineConfig.startIndex &&
                    i <= secondaryLineConfig.endIndex) {
                    pp +=
                        'L' +
                            getX(cumulativeSpacingSecondary, i) +
                            ' ' +
                            getSecondaryY(secondaryData[i].value) +
                            ' ';
                }
            }
            setSecondaryPoints(pp.replace('L', 'M'));
            if (secondaryData.length > 1 && secondaryLineConfig.showArrow) {
                var ppArray = pp.trim().split(' ');
                var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                var y1 = parseInt(ppArray[ppArray.length - 3]);
                var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, (_a = secondaryLineConfig.arrowConfig) === null || _a === void 0 ? void 0 : _a.length, (_b = secondaryLineConfig.arrowConfig) === null || _b === void 0 ? void 0 : _b.width, (_c = secondaryLineConfig.arrowConfig) === null || _c === void 0 ? void 0 : _c.showArrowBase);
                setSecondaryArrowPoints(arrowPoints);
            }
            /** *************************          For Area Chart          *************************/
            if (secondaryLineConfig.areaChart) {
                var ppp = '';
                if (secondaryData.length > 0) {
                    ppp = 'L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                    ppp += pp;
                    ppp +=
                        'L' +
                            getX(cumulativeSpacingSecondary, secondaryData.length - 1) +
                            ' ' +
                            heightUptoXaxis;
                    ppp += ' L' + initialSpacing + ' ' + heightUptoXaxis + ' ';
                    setSecondaryFillPoints(ppp.replace('L', 'M'));
                }
            }
        }
        else {
            /** *************************          For Curved Charts         *************************/
            var p1Array = [];
            for (var i = 0; i < secondaryData.length; i++) {
                if (i >= secondaryLineConfig.startIndex &&
                    i <= secondaryLineConfig.endIndex) {
                    p1Array.push([
                        getX(cumulativeSpacingSecondary, i),
                        getSecondaryY(secondaryData[i].value)
                    ]);
                }
            }
            var xx = svgPath(p1Array, secondaryLineConfig.curveType, secondaryLineConfig.curvature);
            setSecondaryPoints(xx);
            if (secondaryData.length > 1 && ((_d = props.showArrow1) !== null && _d !== void 0 ? _d : props.showArrows)) {
                var arrowTipY = p1Array[p1Array.length - 1][1];
                var arrowTipX = p1Array[p1Array.length - 1][0];
                var y1 = p1Array[p1Array.length - 2][1];
                var x1 = p1Array[p1Array.length - 2][0];
                var arrowPoints = getArrowPoints(arrowTipX, arrowTipY, x1, y1, arrowLength1, arrowWidth1, showArrowBase1);
                setSecondaryArrowPoints(arrowPoints);
            }
            /** *************************          For Curved Area Charts          *************************/
            if (secondaryLineConfig.areaChart) {
                if (secondaryData.length > 0) {
                    xx = addLeadingAndTrailingPathForAreaFill(xx, secondaryData[0].value, secondaryData.length);
                    setSecondaryFillPoints(xx);
                }
            }
        }
    }, [secondaryData, secondaryLineConfig]);
    var gradientDirection = (_261 = props.gradientDirection) !== null && _261 !== void 0 ? _261 : 'vertical';
    var horizSections = [{ value: '0' }];
    var stepHeight = (_262 = props.stepHeight) !== null && _262 !== void 0 ? _262 : containerHeight / noOfSections;
    var stepValue = (_263 = props.stepValue) !== null && _263 !== void 0 ? _263 : maxValue / noOfSections;
    var noOfSectionsBelowXAxis = (_264 = props.noOfSectionsBelowXAxis) !== null && _264 !== void 0 ? _264 : Math.round(Math.ceil(-mostNegativeValue / stepValue));
    var negativeStepValue = (_265 = props.negativeStepValue) !== null && _265 !== void 0 ? _265 : stepValue;
    var axesAndRulesProps = getAxesAndRulesProps(props, containerHeight, stepHeight, stepValue, noOfSections, roundToDigits, negativeStepValue, secondaryMaxValue, secondaryMinItem, showSecondaryFractionalValues, secondaryRoundToDigits);
    var secondaryNoOfSectionsBelowXAxis = axesAndRulesProps.secondaryYAxisConfig.noOfSectionsBelowXAxis;
    var fourthQuadrantHeight = Math.max(noOfSectionsBelowXAxis, secondaryNoOfSectionsBelowXAxis) *
        stepHeight;
    var containerHeightIncludingBelowXAxis = extendedContainerHeight + fourthQuadrantHeight;
    var mostNegativeValueOnYAxis = negativeStepValue * noOfSectionsBelowXAxis;
    var getX = function (spacingArray, index) {
        return initialSpacing + (index ? spacingArray[index - 1] : 0);
    };
    var getY = function (value) {
        if (containsNegativeValue && value < 0 && stepValue !== negativeStepValue) {
            return (extendedContainerHeight +
                (value * fourthQuadrantHeight) / -mostNegativeValueOnYAxis);
        }
        return extendedContainerHeight - (value * containerHeight) / maxValue;
    };
    var showXAxisIndices = (_266 = props.showXAxisIndices) !== null && _266 !== void 0 ? _266 : AxesAndRulesDefaults.showXAxisIndices;
    var xAxisIndicesHeight = (_267 = props.xAxisIndicesHeight) !== null && _267 !== void 0 ? _267 : AxesAndRulesDefaults.xAxisIndicesHeight;
    var xAxisIndicesWidth = (_268 = props.xAxisIndicesWidth) !== null && _268 !== void 0 ? _268 : AxesAndRulesDefaults.xAxisIndicesWidth;
    var xAxisIndicesColor = (_269 = props.xAxisIndicesColor) !== null && _269 !== void 0 ? _269 : AxesAndRulesDefaults.xAxisIndicesColor;
    var xAxisLabelsAtBottom = (_270 = props.xAxisLabelsAtBottom) !== null && _270 !== void 0 ? _270 : false;
    var xAxisTextNumberOfLines = (_271 = props.xAxisTextNumberOfLines) !== null && _271 !== void 0 ? _271 : AxesAndRulesDefaults.xAxisTextNumberOfLines;
    var xAxisLabelsVerticalShift = (_272 = props.xAxisLabelsVerticalShift) !== null && _272 !== void 0 ? _272 : (xAxisLabelsAtBottom
        ? fourthQuadrantHeight
        : AxesAndRulesDefaults.xAxisLabelsVerticalShift);
    var horizontalRulesStyle = props.horizontalRulesStyle;
    var horizontal = false;
    var yAxisAtTop = false;
    defaultPointerConfig.pointerStripHeight =
        containerHeight + noOfSectionsBelowXAxis * stepHeight;
    var pointerConfig = props.pointerConfig;
    var getPointerProps = (_273 = props.getPointerProps) !== null && _273 !== void 0 ? _273 : null;
    var pointerHeight = (_274 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.height) !== null && _274 !== void 0 ? _274 : defaultPointerConfig.height;
    var pointerWidth = (_275 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.width) !== null && _275 !== void 0 ? _275 : defaultPointerConfig.width;
    var pointerRadius = (_276 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.radius) !== null && _276 !== void 0 ? _276 : defaultPointerConfig.radius;
    var pointerColor = (_277 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerColor) !== null && _277 !== void 0 ? _277 : defaultPointerConfig.pointerColor;
    var pointerComponent = (_278 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerComponent) !== null && _278 !== void 0 ? _278 : defaultPointerConfig.pointerComponent;
    var showPointerStrip = (pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.showPointerStrip) === false
        ? false
        : defaultPointerConfig.showPointerStrip;
    var pointerStripHeight = (_279 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripHeight) !== null && _279 !== void 0 ? _279 : defaultPointerConfig.pointerStripHeight;
    var pointerStripWidth = (_280 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripWidth) !== null && _280 !== void 0 ? _280 : defaultPointerConfig.pointerStripWidth;
    var pointerStripColor = (_281 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripColor) !== null && _281 !== void 0 ? _281 : defaultPointerConfig.pointerStripColor;
    var pointerStripUptoDataPoint = (_282 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripUptoDataPoint) !== null && _282 !== void 0 ? _282 : defaultPointerConfig.pointerStripUptoDataPoint;
    var pointerLabelComponent = (_283 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelComponent) !== null && _283 !== void 0 ? _283 : defaultPointerConfig.pointerLabelComponent;
    var stripOverPointer = (_284 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.stripOverPointer) !== null && _284 !== void 0 ? _284 : defaultPointerConfig.stripOverPointer;
    var shiftPointerLabelX = (_285 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.shiftPointerLabelX) !== null && _285 !== void 0 ? _285 : defaultPointerConfig.shiftPointerLabelX;
    var shiftPointerLabelY = (_286 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.shiftPointerLabelY) !== null && _286 !== void 0 ? _286 : defaultPointerConfig.shiftPointerLabelY;
    var pointerLabelWidth = (_287 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelWidth) !== null && _287 !== void 0 ? _287 : defaultPointerConfig.pointerLabelWidth;
    var pointerLabelHeight = (_288 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelHeight) !== null && _288 !== void 0 ? _288 : defaultPointerConfig.pointerLabelHeight;
    var autoAdjustPointerLabelPosition = (_289 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.autoAdjustPointerLabelPosition) !== null && _289 !== void 0 ? _289 : defaultPointerConfig.autoAdjustPointerLabelPosition;
    var pointerVanishDelay = (_290 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerVanishDelay) !== null && _290 !== void 0 ? _290 : defaultPointerConfig.pointerVanishDelay;
    var activatePointersOnLongPress = (_291 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.activatePointersOnLongPress) !== null && _291 !== void 0 ? _291 : defaultPointerConfig.activatePointersOnLongPress;
    var activatePointersInstantlyOnTouch = (_292 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.activatePointersInstantlyOnTouch) !== null && _292 !== void 0 ? _292 : defaultPointerConfig.activatePointersInstantlyOnTouch;
    var activatePointersDelay = (_293 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.activatePointersDelay) !== null && _293 !== void 0 ? _293 : defaultPointerConfig.activatePointersDelay;
    var initialPointerIndex = (_294 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.initialPointerIndex) !== null && _294 !== void 0 ? _294 : defaultPointerConfig.initialPointerIndex;
    var initialPointerAppearDelay = (_295 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.initialPointerAppearDelay) !== null && _295 !== void 0 ? _295 : (isAnimated
        ? animationDuration
        : defaultPointerConfig.initialPointerAppearDelay);
    var persistPointer = (_296 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.persistPointer) !== null && _296 !== void 0 ? _296 : defaultPointerConfig.persistPointer;
    var resetPointerIndexOnRelease = (_297 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.resetPointerIndexOnRelease) !== null && _297 !== void 0 ? _297 : defaultPointerConfig.resetPointerIndexOnRelease;
    var hidePointers = (_298 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointers) !== null && _298 !== void 0 ? _298 : defaultPointerConfig.hidePointers;
    var hidePointer1 = (_299 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer1) !== null && _299 !== void 0 ? _299 : defaultPointerConfig.hidePointer1;
    var hidePointer2 = (_300 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer2) !== null && _300 !== void 0 ? _300 : defaultPointerConfig.hidePointer2;
    var hidePointer3 = (_301 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer3) !== null && _301 !== void 0 ? _301 : defaultPointerConfig.hidePointer3;
    var hidePointer4 = (_302 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer4) !== null && _302 !== void 0 ? _302 : defaultPointerConfig.hidePointer4;
    var hidePointer5 = (_303 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer5) !== null && _303 !== void 0 ? _303 : defaultPointerConfig.hidePointer5;
    var hideSecondaryPointer = (_304 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hideSecondaryPointer) !== null && _304 !== void 0 ? _304 : defaultPointerConfig.hideSecondaryPointer;
    var resetPointerOnDataChange = (_305 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.resetPointerOnDataChange) !== null && _305 !== void 0 ? _305 : defaultPointerConfig.resetPointerOnDataChange;
    var hidePointerDataPointForMissingValues = (_306 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointerDataPointForMissingValues) !== null && _306 !== void 0 ? _306 : ((pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointerForMissingValues) ? true : false);
    var pointerEvents = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerEvents;
    var disableScroll = (_307 = props.disableScroll) !== null && _307 !== void 0 ? _307 : (pointerConfig
        ? activatePointersOnLongPress
            ? !!responderActive
            : true
        : false);
    var showScrollIndicator = (_308 = props.showScrollIndicator) !== null && _308 !== void 0 ? _308 : LineDefaults.showScrollIndicator;
    var focusEnabled = (_309 = props.focusEnabled) !== null && _309 !== void 0 ? _309 : LineDefaults.focusEnabled;
    var showDataPointOnFocus = (_310 = props.showDataPointOnFocus) !== null && _310 !== void 0 ? _310 : LineDefaults.showDataPointOnFocus;
    var showStripOnFocus = (_311 = props.showStripOnFocus) !== null && _311 !== void 0 ? _311 : LineDefaults.showStripOnFocus;
    var showTextOnFocus = (_312 = props.showTextOnFocus) !== null && _312 !== void 0 ? _312 : LineDefaults.showTextOnFocus;
    var showDataPointLabelOnFocus = (_313 = props.showDataPointLabelOnFocus) !== null && _313 !== void 0 ? _313 : LineDefaults.showDataPointLabelOnFocus;
    var stripHeight = props.stripHeight;
    var stripWidth = (_314 = props.stripWidth) !== null && _314 !== void 0 ? _314 : LineDefaults.stripWidth;
    var stripColor = (_315 = props.stripColor) !== null && _315 !== void 0 ? _315 : color1;
    var stripOpacity = (_316 = props.stripOpacity) !== null && _316 !== void 0 ? _316 : (startOpacity1 + endOpacity1) / 2;
    var stripStrokeDashArray = props.stripStrokeDashArray;
    var unFocusOnPressOut = (_317 = props.unFocusOnPressOut) !== null && _317 !== void 0 ? _317 : LineDefaults.unFocusOnPressOut;
    var delayBeforeUnFocus = (_318 = props.delayBeforeUnFocus) !== null && _318 !== void 0 ? _318 : LineDefaults.delayBeforeUnFocus;
    var focusTogether = (_319 = props.focusTogether) !== null && _319 !== void 0 ? _319 : true;
    var focusProximity = (_320 = props.focusProximity) !== null && _320 !== void 0 ? _320 : Infinity;
    var lineGradient = (_321 = props.lineGradient) !== null && _321 !== void 0 ? _321 : LineDefaults.lineGradient;
    var lineGradientDirection = (_322 = props.lineGradientDirection) !== null && _322 !== void 0 ? _322 : 'vertical';
    var lineGradientStartColor = (_323 = props.lineGradientStartColor) !== null && _323 !== void 0 ? _323 : LineDefaults.lineGradientStartColor;
    var lineGradientEndColor = (_324 = props.lineGradientEndColor) !== null && _324 !== void 0 ? _324 : LineDefaults.lineGradientEndColor;
    var _393 = __read(useState(-1), 2), selectedLineNumber = _393[0], setSelectedLineNumber = _393[1];
    var getPointerY = function (value) {
        return value || value === 0
            ? containerHeight -
                (value * containerHeight) / maxValue -
                (pointerRadius || pointerHeight / 2) +
                10
            : 0;
    };
    var initialisePointers = function () {
        var _a, _b, _c, _d;
        if (initialPointerIndex !== -1) {
            var x_1 = initialSpacing +
                spacing * initialPointerIndex -
                (pointerRadius || pointerWidth / 2) -
                1;
            if (dataSet === null || dataSet === void 0 ? void 0 : dataSet.length) {
                var item_1 = dataSet[0].data[initialPointerIndex];
                if (initialPointerAppearDelay) {
                    setTimeout(function () {
                        setPointerConfigForDataSet(initialPointerIndex, item_1, x_1);
                    }, initialPointerAppearDelay);
                }
                else {
                    setPointerConfigForDataSet(initialPointerIndex, item_1, x_1);
                }
            }
            else {
                var item_2 = (data0 !== null && data0 !== void 0 ? data0 : data)[initialPointerIndex];
                var y_1 = getPointerY(item_2.value);
                var y2_1 = getPointerY((_a = data2 === null || data2 === void 0 ? void 0 : data2[initialPointerIndex]) === null || _a === void 0 ? void 0 : _a.value);
                var y3_1 = getPointerY((_b = data3 === null || data3 === void 0 ? void 0 : data3[initialPointerIndex]) === null || _b === void 0 ? void 0 : _b.value);
                var y4_1 = getPointerY((_c = data4 === null || data4 === void 0 ? void 0 : data4[initialPointerIndex]) === null || _c === void 0 ? void 0 : _c.value);
                var y5_1 = getPointerY((_d = data5 === null || data5 === void 0 ? void 0 : data5[initialPointerIndex]) === null || _d === void 0 ? void 0 : _d.value);
                if (initialPointerAppearDelay) {
                    setTimeout(function () {
                        setPointerConfig(initialPointerIndex, item_2, x_1, y_1, y2_1, y3_1, y4_1, y5_1);
                    }, initialPointerAppearDelay);
                }
                else {
                    setPointerConfig(initialPointerIndex, item_2, x_1, y_1, y2_1, y3_1, y4_1, y5_1);
                }
            }
        }
    };
    useEffect(function () {
        initialisePointers();
    }, []);
    useEffect(function () {
        if (resetPointerOnDataChange) {
            initialisePointers();
        }
    }, [data]);
    var setPointerConfig = function (initialPointerIndex, item, x, y, y2, y3, y4, y5) {
        setPointerIndex(initialPointerIndex);
        setPointerItem(item);
        setPointerX(x);
        setPointerY(y);
        setPointerY2(y2);
        setPointerY3(y3);
        setPointerY4(y4);
        setPointerY5(y5);
    };
    var setPointerConfigForDataSet = function (initialPointerIndex, item, x) {
        var _a;
        setPointerIndex(initialPointerIndex);
        setPointerItem(item);
        setPointerX(x);
        var initialPointerYs = (_a = dataSet === null || dataSet === void 0 ? void 0 : dataSet.map(function (set) { return getPointerY(set.data[initialPointerIndex].value); })) !== null && _a !== void 0 ? _a : [];
        setPointerYsForDataSet(initialPointerYs);
    };
    var handleFocus = function (index, item, locationY, onStripPress) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        var lineNumber = 0;
        var closestItem = item;
        if (dataSet) {
            var minDistance_1 = Infinity;
            dataSet.forEach(function (setItem, setIndex) {
                var _a;
                var distance = Math.abs(getY((_a = setItem.data[index]) === null || _a === void 0 ? void 0 : _a.value) - locationY);
                if (distance < minDistance_1) {
                    minDistance_1 = distance;
                    lineNumber = setIndex + 1;
                    closestItem = setItem.data[index];
                }
            });
        }
        else {
            var distance1 = void 0, distance2 = void 0, distance3 = void 0, distance4 = void 0, distance5 = void 0, distance6 = void 0;
            var minDistance = Infinity;
            if (typeof ((_a = data[index]) === null || _a === void 0 ? void 0 : _a.value) === 'number') {
                distance1 = Math.abs(getY((_b = data[index]) === null || _b === void 0 ? void 0 : _b.value) - locationY);
                minDistance = distance1;
                if (distance1 < focusProximity) {
                    lineNumber = 1;
                    closestItem = data[index];
                }
            }
            if (typeof ((_c = data2[index]) === null || _c === void 0 ? void 0 : _c.value) === 'number') {
                distance2 = Math.abs(getY((_d = data2[index]) === null || _d === void 0 ? void 0 : _d.value) - locationY);
                if (minDistance > distance2 && distance2 < focusProximity) {
                    minDistance = distance2;
                    lineNumber = 2;
                    closestItem = data2[index];
                }
            }
            if (typeof ((_e = data3[index]) === null || _e === void 0 ? void 0 : _e.value) === 'number') {
                distance3 = Math.abs(getY((_f = data3[index]) === null || _f === void 0 ? void 0 : _f.value) - locationY);
                if (minDistance > distance3 && distance3 < focusProximity) {
                    minDistance = distance3;
                    lineNumber = 3;
                    closestItem = data3[index];
                }
            }
            if (typeof ((_g = data4[index]) === null || _g === void 0 ? void 0 : _g.value) === 'number') {
                distance4 = Math.abs(getY((_h = data4[index]) === null || _h === void 0 ? void 0 : _h.value) - locationY);
                if (minDistance > distance4 && distance4 < focusProximity) {
                    minDistance = distance4;
                    lineNumber = 4;
                    closestItem = data4[index];
                }
            }
            if (typeof ((_j = data5[index]) === null || _j === void 0 ? void 0 : _j.value) === 'number') {
                distance5 = Math.abs(getY((_k = data5[index]) === null || _k === void 0 ? void 0 : _k.value) - locationY);
                if (minDistance > distance5 && distance5 < focusProximity) {
                    minDistance = distance5;
                    lineNumber = 5;
                    closestItem = data5[index];
                }
            }
            if (typeof ((_l = secondaryData[index]) === null || _l === void 0 ? void 0 : _l.value) === 'number') {
                distance6 = Math.abs(getY((_m = secondaryData[index]) === null || _m === void 0 ? void 0 : _m.value) - locationY);
                if (minDistance > distance6 && distance6 < focusProximity) {
                    minDistance = distance6;
                    lineNumber = 6666;
                    closestItem = secondaryData[index];
                }
            }
        }
        setSelectedLineNumber(lineNumber - 1);
        if (lineNumber) {
            onStripPress(closestItem, index);
        }
    };
    var handleUnFocus = function () {
        if (unFocusOnPressOut) {
            setTimeout(function () { return setSelectedIndex(-1); }, delayBeforeUnFocus);
        }
    };
    var dataPointsRadius = (_326 = (_325 = props.dataPointsRadius1) !== null && _325 !== void 0 ? _325 : props.dataPointsRadius) !== null && _326 !== void 0 ? _326 : LineDefaults.dataPointsRadius;
    var dataPointsWidth = (_328 = (_327 = props.dataPointsWidth1) !== null && _327 !== void 0 ? _327 : props.dataPointsWidth) !== null && _328 !== void 0 ? _328 : LineDefaults.dataPointsWidth;
    var extraWidthDueToDataPoint = props.hideDataPoints
        ? 0
        : dataPointsRadius !== null && dataPointsRadius !== void 0 ? dataPointsRadius : dataPointsWidth;
    var barAndLineChartsWrapperProps = {
        chartType: chartTypes.LINE,
        containerHeight: containerHeight,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: stepHeight,
        negativeStepHeight: stepHeight,
        labelsExtraHeight: labelsExtraHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        rtl: false,
        shiftX: 0,
        shiftY: 0,
        yAxisAtTop: yAxisAtTop,
        initialSpacing: initialSpacing,
        data: data0 !== null && data0 !== void 0 ? data0 : data,
        stackData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        secondaryData: secondaryData,
        barWidth: 0, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        xAxisThickness: xAxisThickness,
        totalWidth: totalWidth,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollToIndex: props.scrollToIndex,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        indicatorColor: props.indicatorColor,
        selectedIndex: [selectedIndex],
        setSelectedIndex: setSelectedIndex,
        spacing: spacing,
        showLine: false,
        lineConfig: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineConfig2: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        maxValue: maxValue,
        lineData: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineData2: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineBehindBars: false,
        points: points,
        points2: '', // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        arrowPoints: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        // horizSectionProps-
        width: props.width,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        sectionColors: props.sectionColors,
        showFractionalValues: showFractionalValues,
        axesAndRulesProps: axesAndRulesProps,
        yAxisLabelTexts: props.yAxisLabelTexts,
        yAxisOffset: yAxisOffset,
        rotateYAxisTexts: 0,
        hideAxesAndRules: props.hideAxesAndRules,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        pointerConfig: pointerConfig,
        getPointerProps: getPointerProps,
        pointerIndex: pointerIndex,
        pointerX: pointerX,
        pointerY: pointerY,
        onEndReached: props.onEndReached,
        onStartReached: props.onStartReached,
        endReachedOffset: (_329 = props.endReachedOffset) !== null && _329 !== void 0 ? _329 : LineDefaults.endReachedOffset,
        onMomentumScrollEnd: props.onMomentumScrollEnd,
        extraWidthDueToDataPoint: extraWidthDueToDataPoint,
        customBackground: props.customBackground,
        onlyPositive: onlyPositive,
        highlightEnabled: LineDefaults.highlightEnabled,
        lowlightOpacity: LineDefaults.lowlightOpacity,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        onScrollEndDrag: props.onScrollEndDrag,
        allowFontScaling: allowFontScaling
    };
    var pointerItemLocal = [];
    if (pointerConfig) {
        pointerItemLocal = [
            __assign(__assign({}, pointerItem), { value: (_331 = (_330 = props.data) === null || _330 === void 0 ? void 0 : _330[pointerIndex]) === null || _331 === void 0 ? void 0 : _331.value })
        ];
        if (pointerY2 !== 0) {
            pointerItemLocal.push(__assign(__assign({}, pointerItem), { value: (_333 = (_332 = props.data2) === null || _332 === void 0 ? void 0 : _332[pointerIndex]) === null || _333 === void 0 ? void 0 : _333.value }));
        }
        if (pointerY3 !== 0) {
            pointerItemLocal.push(__assign(__assign({}, pointerItem), { value: (_335 = (_334 = props.data3) === null || _334 === void 0 ? void 0 : _334[pointerIndex]) === null || _335 === void 0 ? void 0 : _335.value }));
        }
        if (pointerY4 !== 0) {
            pointerItemLocal.push(__assign(__assign({}, pointerItem), { value: (_337 = (_336 = props.data4) === null || _336 === void 0 ? void 0 : _336[pointerIndex]) === null || _337 === void 0 ? void 0 : _337.value }));
        }
        if (pointerY5 !== 0) {
            pointerItemLocal.push(__assign(__assign({}, pointerItem), { value: (_339 = (_338 = props.data5) === null || _338 === void 0 ? void 0 : _338[pointerIndex]) === null || _339 === void 0 ? void 0 : _339.value }));
        }
    }
    return {
        curvature: curvature,
        curveType: curveType,
        scrollX: scrollX,
        setScrollX: setScrollX,
        arrow1Points: arrow1Points,
        setArrow1Points: setArrow1Points,
        arrow2Points: arrow2Points,
        setArrow2Points: setArrow2Points,
        arrow3Points: arrow3Points,
        setArrow3Points: setArrow3Points,
        arrow4Points: arrow4Points,
        setArrow4Points: setArrow4Points,
        arrow5Points: arrow5Points,
        setArrow5Points: setArrow5Points,
        secondaryArrowPoints: secondaryArrowPoints,
        setSecondaryArrowPoints: setSecondaryArrowPoints,
        pointerIndex: pointerIndex,
        setPointerIndex: setPointerIndex,
        pointerX: pointerX,
        setPointerX: setPointerX,
        pointerY: pointerY,
        setPointerY: setPointerY,
        pointerItem: pointerItem,
        setPointerItem: setPointerItem,
        pointerY2: pointerY2,
        setPointerY2: setPointerY2,
        pointerItem2: pointerItem2,
        setPointerItem2: setPointerItem2,
        pointerY3: pointerY3,
        setPointerY3: setPointerY3,
        pointerItem3: pointerItem3,
        setPointerItem3: setPointerItem3,
        pointerY4: pointerY4,
        setPointerY4: setPointerY4,
        pointerItem4: pointerItem4,
        setPointerItem4: setPointerItem4,
        pointerY5: pointerY5,
        setPointerY5: setPointerY5,
        pointerYsForDataSet: pointerYsForDataSet,
        setPointerYsForDataSet: setPointerYsForDataSet,
        pointerItem5: pointerItem5,
        setPointerItem5: setPointerItem5,
        secondaryPointerY: secondaryPointerY,
        setSecondaryPointerY: setSecondaryPointerY,
        secondaryPointerItem: secondaryPointerItem,
        setSecondaryPointerItem: setSecondaryPointerItem,
        pointerItemsForSet: pointerItemsForSet,
        setPointerItemsForSet: setPointerItemsForSet,
        secondaryPointerItemsForSet: secondaryPointerItemsForSet,
        setSecondaryPointerItemsForSet: setSecondaryPointerItemsForSet,
        responderStartTime: responderStartTime,
        setResponderStartTime: setResponderStartTime,
        responderActive: responderActive,
        setResponderActive: setResponderActive,
        points: points,
        setPoints: setPoints,
        points2: points2,
        setPoints2: setPoints2,
        points3: points3,
        setPoints3: setPoints3,
        points4: points4,
        setPoints4: setPoints4,
        points5: points5,
        setPoints5: setPoints5,
        secondaryPoints: secondaryPoints,
        setSecondaryPoints: setSecondaryPoints,
        fillPoints: fillPoints,
        setFillPoints: setFillPoints,
        fillPoints2: fillPoints2,
        setFillPoints2: setFillPoints2,
        fillPoints3: fillPoints3,
        setFillPoints3: setFillPoints3,
        fillPoints4: fillPoints4,
        setFillPoints4: setFillPoints4,
        fillPoints5: fillPoints5,
        setFillPoints5: setFillPoints5,
        secondaryFillPoints: secondaryFillPoints,
        setSecondaryFillPoints: setSecondaryFillPoints,
        pointsFromSet: pointsFromSet,
        setPointsFromSet: setPointsFromSet,
        fillPointsFromSet: fillPointsFromSet,
        setFillPointsFromSet: setFillPointsFromSet,
        arrowPointsFromSet: arrowPointsFromSet,
        setArrowPointsFromSet: setArrowPointsFromSet,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        noOfSections: noOfSections,
        containerHeight: containerHeight,
        data: data,
        data2: data2,
        data3: data3,
        data4: data4,
        data5: data5,
        secondaryData: secondaryData,
        dataSet: dataSet,
        data0: data0,
        scrollToEnd: scrollToEnd,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        labelsExtraHeight: labelsExtraHeight,
        animationDuration: animationDuration,
        onDataChangeAnimationDuration: onDataChangeAnimationDuration,
        animateTogether: animateTogether,
        renderDataPointsAfterAnimationEnds: renderDataPointsAfterAnimationEnds,
        animateOnDataChange: animateOnDataChange,
        startIndex1: startIndex1,
        startIndex2: startIndex2,
        endIndex1: endIndex1,
        endIndex2: endIndex2,
        startIndex3: startIndex3,
        endIndex3: endIndex3,
        startIndex4: startIndex4,
        endIndex4: endIndex4,
        startIndex5: startIndex5,
        endIndex5: endIndex5,
        lineSegments: lineSegments,
        lineSegments2: lineSegments2,
        lineSegments3: lineSegments3,
        lineSegments4: lineSegments4,
        lineSegments5: lineSegments5,
        highlightedRange: highlightedRange,
        adjustToWidth: adjustToWidth,
        initialSpacing: initialSpacing,
        endSpacing: endSpacing,
        thickness: thickness,
        yAxisLabelWidth: yAxisLabelWidth,
        spacing: spacing,
        xAxisThickness: xAxisThickness,
        dataPointsHeight1: dataPointsHeight1,
        dataPointsWidth1: dataPointsWidth1,
        dataPointsRadius1: dataPointsRadius1,
        dataPointsColor1: dataPointsColor1,
        dataPointsShape1: dataPointsShape1,
        dataPointsHeight2: dataPointsHeight2,
        dataPointsWidth2: dataPointsWidth2,
        dataPointsRadius2: dataPointsRadius2,
        dataPointsColor2: dataPointsColor2,
        dataPointsShape2: dataPointsShape2,
        dataPointsHeight3: dataPointsHeight3,
        dataPointsWidth3: dataPointsWidth3,
        dataPointsRadius3: dataPointsRadius3,
        dataPointsColor3: dataPointsColor3,
        dataPointsShape3: dataPointsShape3,
        dataPointsHeight4: dataPointsHeight4,
        dataPointsWidth4: dataPointsWidth4,
        dataPointsRadius4: dataPointsRadius4,
        dataPointsColor4: dataPointsColor4,
        dataPointsShape4: dataPointsShape4,
        dataPointsHeight5: dataPointsHeight5,
        dataPointsWidth5: dataPointsWidth5,
        dataPointsRadius5: dataPointsRadius5,
        dataPointsColor5: dataPointsColor5,
        dataPointsShape5: dataPointsShape5,
        areaChart: areaChart,
        areaChart1: areaChart1,
        areaChart2: areaChart2,
        areaChart3: areaChart3,
        areaChart4: areaChart4,
        areaChart5: areaChart5,
        atLeastOneAreaChart: atLeastOneAreaChart,
        getIsNthAreaChart: getIsNthAreaChart,
        stepChart: stepChart,
        stepChart1: stepChart1,
        stepChart2: stepChart2,
        stepChart3: stepChart3,
        stepChart4: stepChart4,
        stepChart5: stepChart5,
        edgePosition: edgePosition,
        textFontSize1: textFontSize1,
        textFontSize2: textFontSize2,
        textFontSize3: textFontSize3,
        textFontSize4: textFontSize4,
        textFontSize5: textFontSize5,
        textColor1: textColor1,
        textColor2: textColor2,
        textColor3: textColor3,
        textColor4: textColor4,
        textColor5: textColor5,
        totalWidth: totalWidth,
        maxValue: maxValue,
        mostNegativeValue: mostNegativeValue,
        overflowTop: overflowTop,
        extendedContainerHeight: extendedContainerHeight,
        getX: getX,
        getY: getY,
        secondaryMaxValue: secondaryMaxValue,
        getSecondaryY: getSecondaryY,
        heightUptoXaxis: heightUptoXaxis,
        showValuesAsDataPointsText: showValuesAsDataPointsText,
        thickness1: thickness1,
        thickness2: thickness2,
        thickness3: thickness3,
        thickness4: thickness4,
        thickness5: thickness5,
        zIndex1: zIndex1,
        zIndex2: zIndex2,
        zIndex3: zIndex3,
        zIndex4: zIndex4,
        zIndex5: zIndex5,
        strokeDashArray1: strokeDashArray1,
        strokeDashArray2: strokeDashArray2,
        strokeDashArray3: strokeDashArray3,
        strokeDashArray4: strokeDashArray4,
        strokeDashArray5: strokeDashArray5,
        strokeLinecap1: strokeLinecap1,
        strokeLinecap2: strokeLinecap2,
        strokeLinecap3: strokeLinecap3,
        strokeLinecap4: strokeLinecap4,
        strokeLinecap5: strokeLinecap5,
        rotateLabel: rotateLabel,
        isAnimated: isAnimated,
        hidePointers: hidePointers,
        hideDataPoints1: hideDataPoints1,
        hideDataPoints2: hideDataPoints2,
        hideDataPoints3: hideDataPoints3,
        hideDataPoints4: hideDataPoints4,
        hideDataPoints5: hideDataPoints5,
        color1: color1,
        color2: color2,
        color3: color3,
        color4: color4,
        color5: color5,
        startFillColor1: startFillColor1,
        endFillColor1: endFillColor1,
        startOpacity: startOpacity,
        endOpacity: endOpacity,
        startOpacity1: startOpacity1,
        endOpacity1: endOpacity1,
        startFillColor2: startFillColor2,
        endFillColor2: endFillColor2,
        startOpacity2: startOpacity2,
        endOpacity2: endOpacity2,
        startFillColor3: startFillColor3,
        endFillColor3: endFillColor3,
        startOpacity3: startOpacity3,
        endOpacity3: endOpacity3,
        startFillColor4: startFillColor4,
        endFillColor4: endFillColor4,
        startOpacity4: startOpacity4,
        endOpacity4: endOpacity4,
        startFillColor5: startFillColor5,
        endFillColor5: endFillColor5,
        startOpacity5: startOpacity5,
        endOpacity5: endOpacity5,
        arrowLength1: arrowLength1,
        arrowWidth1: arrowWidth1,
        arrowStrokeWidth1: arrowStrokeWidth1,
        arrowStrokeColor1: arrowStrokeColor1,
        arrowFillColor1: arrowFillColor1,
        showArrowBase1: showArrowBase1,
        arrowLength2: arrowLength2,
        arrowWidth2: arrowWidth2,
        arrowStrokeWidth2: arrowStrokeWidth2,
        arrowStrokeColor2: arrowStrokeColor2,
        arrowFillColor2: arrowFillColor2,
        showArrowBase2: showArrowBase2,
        arrowLength3: arrowLength3,
        arrowWidth3: arrowWidth3,
        arrowStrokeWidth3: arrowStrokeWidth3,
        arrowStrokeColor3: arrowStrokeColor3,
        arrowFillColor3: arrowFillColor3,
        showArrowBase3: showArrowBase3,
        arrowLength4: arrowLength4,
        arrowWidth4: arrowWidth4,
        arrowStrokeWidth4: arrowStrokeWidth4,
        arrowStrokeColor4: arrowStrokeColor4,
        arrowFillColor4: arrowFillColor4,
        showArrowBase4: showArrowBase4,
        arrowLength5: arrowLength5,
        arrowWidth5: arrowWidth5,
        arrowStrokeWidth5: arrowStrokeWidth5,
        arrowStrokeColor5: arrowStrokeColor5,
        arrowFillColor5: arrowFillColor5,
        showArrowBase5: showArrowBase5,
        arrowLengthsFromSet: arrowLengthsFromSet,
        arrowWidthsFromSet: arrowWidthsFromSet,
        arrowStrokeWidthsFromSet: arrowStrokeWidthsFromSet,
        arrowStrokeColorsFromSet: arrowStrokeColorsFromSet,
        arrowFillColorsFromSet: arrowFillColorsFromSet,
        showArrowBasesFromSet: showArrowBasesFromSet,
        secondaryLineConfig: secondaryLineConfig,
        addLeadingAndTrailingPathForAreaFill: addLeadingAndTrailingPathForAreaFill,
        getNextPoint: getNextPoint,
        getStepPath: getStepPath,
        getSegmentPath: getSegmentPath,
        gradientDirection: gradientDirection,
        horizSections: horizSections,
        stepHeight: stepHeight,
        stepValue: stepValue,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        horizontalRulesStyle: horizontalRulesStyle,
        showFractionalValues: showFractionalValues,
        roundToDigits: roundToDigits,
        horizontal: horizontal,
        yAxisAtTop: yAxisAtTop,
        pointerConfig: pointerConfig,
        getPointerProps: getPointerProps,
        pointerHeight: pointerHeight,
        pointerWidth: pointerWidth,
        pointerRadius: pointerRadius,
        pointerColor: pointerColor,
        pointerComponent: pointerComponent,
        showPointerStrip: showPointerStrip,
        pointerStripHeight: pointerStripHeight,
        pointerStripWidth: pointerStripWidth,
        pointerStripColor: pointerStripColor,
        pointerStripUptoDataPoint: pointerStripUptoDataPoint,
        pointerLabelComponent: pointerLabelComponent,
        stripOverPointer: stripOverPointer,
        shiftPointerLabelX: shiftPointerLabelX,
        shiftPointerLabelY: shiftPointerLabelY,
        pointerLabelWidth: pointerLabelWidth,
        pointerLabelHeight: pointerLabelHeight,
        autoAdjustPointerLabelPosition: autoAdjustPointerLabelPosition,
        pointerVanishDelay: pointerVanishDelay,
        activatePointersOnLongPress: activatePointersOnLongPress,
        activatePointersInstantlyOnTouch: activatePointersInstantlyOnTouch,
        activatePointersDelay: activatePointersDelay,
        initialPointerIndex: initialPointerIndex,
        initialPointerAppearDelay: initialPointerAppearDelay,
        persistPointer: persistPointer,
        resetPointerIndexOnRelease: resetPointerIndexOnRelease,
        hidePointer1: hidePointer1,
        hidePointer2: hidePointer2,
        hidePointer3: hidePointer3,
        hidePointer4: hidePointer4,
        hidePointer5: hidePointer5,
        hideSecondaryPointer: hideSecondaryPointer,
        resetPointerOnDataChange: resetPointerOnDataChange,
        hidePointerDataPointForMissingValues: hidePointerDataPointForMissingValues,
        pointerEvents: pointerEvents,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        focusEnabled: focusEnabled,
        showDataPointOnFocus: showDataPointOnFocus,
        showStripOnFocus: showStripOnFocus,
        showTextOnFocus: showTextOnFocus,
        showDataPointLabelOnFocus: showDataPointLabelOnFocus,
        stripHeight: stripHeight,
        stripWidth: stripWidth,
        stripColor: stripColor,
        stripOpacity: stripOpacity,
        stripStrokeDashArray: stripStrokeDashArray,
        unFocusOnPressOut: unFocusOnPressOut,
        delayBeforeUnFocus: delayBeforeUnFocus,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        lineGradient: lineGradient,
        lineGradientDirection: lineGradientDirection,
        lineGradientStartColor: lineGradientStartColor,
        lineGradientEndColor: lineGradientEndColor,
        getPointerY: getPointerY,
        initialisePointers: initialisePointers,
        barAndLineChartsWrapperProps: barAndLineChartsWrapperProps,
        yAxisExtraHeightAtTop: yAxisExtraHeightAtTop,
        cumulativeSpacing1: cumulativeSpacing1,
        cumulativeSpacing2: cumulativeSpacing2,
        cumulativeSpacing3: cumulativeSpacing3,
        cumulativeSpacing4: cumulativeSpacing4,
        cumulativeSpacing5: cumulativeSpacing5,
        cumulativeSpacingSecondary: cumulativeSpacingSecondary,
        cumulativeSpacingForSet: cumulativeSpacingForSet,
        stripOverDataPoints: props.stripOverDataPoints,
        strips: strips,
        selectedLineNumber: selectedLineNumber,
        setSelectedLineNumber: setSelectedLineNumber,
        lastLineNumber: lastLineNumber,
        focusTogether: focusTogether,
        focusProximity: focusProximity,
        handleFocus: handleFocus,
        handleUnFocus: handleUnFocus,
        renderTooltip: renderTooltip,
        renderTooltip1: renderTooltip1,
        renderTooltip2: renderTooltip2,
        renderTooltip3: renderTooltip3,
        renderTooltip4: renderTooltip4,
        renderTooltip5: renderTooltip5,
        renderTooltipSecondary: renderTooltipSecondary,
        pointerItemLocal: pointerItemLocal,
        allowFontScaling: allowFontScaling
        // oldPoints
    };
};
