// gifted-charts-core/dist/BarChart/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { useEffect, useMemo, useState } from 'react';
import { getArrowPoints, getAxesAndRulesProps, getExtendedContainerHeightWithPadding, getLineConfigForBarChart, getMaxValue, getMostNegativeValue, getNoOfSections, getXForLineInBar, getYForLineInBar, indexOfFirstNonZeroDigit, maxAndMinUtil, svgPath } from '../utils';
import { AxesAndRulesDefaults, BarDefaults, chartTypes, defaultLineConfig, defaultPointerConfig } from '../utils/constants';
export var useBarChart = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66, _67, _68, _69;
    var heightValue = props.heightValue, widthValue = props.widthValue, opacityValue = props.opacityValue, yAxisOffset = props.yAxisOffset, adjustToWidth = props.adjustToWidth, parentWidth = props.parentWidth, labelsDistanceFromXaxis = props.labelsDistanceFromXaxis, autoShiftLabelsForNegativeStacks = props.autoShiftLabelsForNegativeStacks, focusedBarIndex = props.focusedBarIndex, negativeStepValue = props.negativeStepValue, autoCenterTooltip = props.autoCenterTooltip, floatingYAxisLabels = props.floatingYAxisLabels;
    var allowFontScaling = (_a = props.allowFontScaling) !== null && _a !== void 0 ? _a : AxesAndRulesDefaults.allowFontScaling;
    var _70 = __read(useState(''), 2), points = _70[0], setPoints = _70[1];
    var _71 = __read(useState(''), 2), points2 = _71[0], setPoints2 = _71[1];
    var _72 = __read(useState(''), 2), arrowPoints = _72[0], setArrowPoints = _72[1];
    var _73 = __read(useState(function () {
        if (Array.isArray(focusedBarIndex)) {
            return focusedBarIndex;
        }
        return [focusedBarIndex !== null && focusedBarIndex !== void 0 ? focusedBarIndex : -1];
    }), 2), selectedIndex = _73[0], setSelectedIndex = _73[1];
    var _74 = __read(useState((_b = props.highlightedStackIndex) !== null && _b !== void 0 ? _b : -1), 2), selectedStackIndex = _74[0], setSelectedStackIndex = _74[1];
    var showLine = (_c = props.showLine) !== null && _c !== void 0 ? _c : BarDefaults.showLine;
    useEffect(function () {
        var newIndex = Array.isArray(focusedBarIndex)
            ? focusedBarIndex
            : [focusedBarIndex !== null && focusedBarIndex !== void 0 ? focusedBarIndex : -1];
        setSelectedIndex(newIndex);
    }, [focusedBarIndex]);
    useEffect(function () {
        var _a;
        setSelectedStackIndex((_a = props.highlightedStackIndex) !== null && _a !== void 0 ? _a : -1);
    }, [props.highlightedStackIndex]);
    var highlightEnabled = (_d = props.highlightEnabled) !== null && _d !== void 0 ? _d : BarDefaults.highlightEnabled;
    var highlightedBarIndex = (_e = props.highlightedBarIndex) !== null && _e !== void 0 ? _e : selectedIndex;
    var lowlightOpacity = (_f = props.lowlightOpacity) !== null && _f !== void 0 ? _f : BarDefaults.lowlightOpacity;
    var stackHighlightEnabled = (_g = props.stackHighlightEnabled) !== null && _g !== void 0 ? _g : BarDefaults.stackHighlightEnabled;
    var data = useMemo(function () {
        if (!props.data) {
            return [];
        }
        if (yAxisOffset) {
            return props.data.map(function (item) {
                var _a;
                return (__assign(__assign({}, item), { value: ((_a = item.value) !== null && _a !== void 0 ? _a : 0) - yAxisOffset }));
            });
        }
        return props.data.map(function (item) {
            var _a;
            return (__assign(__assign({}, item), { value: (_a = item.value) !== null && _a !== void 0 ? _a : 0 }));
        });
    }, [yAxisOffset, props.data]);
    var stackData = useMemo(function () {
        if (!props.stackData) {
            return undefined;
        }
        if (yAxisOffset) {
            return props.stackData.map(function (item) {
                var cumulativeSum = 0;
                return __assign(__assign({}, item), { stacks: item.stacks.map(function (stackItem) {
                        var _a;
                        var stack = __assign(__assign({}, stackItem), { value: Math.max(
                            // yAxisOffset is reduced from stackItems as long as their cumulative sum is less than yAxisOffset
                            ((_a = stackItem.value) !== null && _a !== void 0 ? _a : 0) -
                                (cumulativeSum < yAxisOffset
                                    ? yAxisOffset - cumulativeSum
                                    : 0), 0) });
                        cumulativeSum += stackItem.value;
                        return stack;
                    }) });
            });
        }
        return props.stackData;
    }, [yAxisOffset, props.stackData]);
    // adjustToWidth should work for data or stacked data
    var dataLength = (_l = (_j = (_h = props.data) === null || _h === void 0 ? void 0 : _h.length) !== null && _j !== void 0 ? _j : (_k = props.stackData) === null || _k === void 0 ? void 0 : _k.length) !== null && _l !== void 0 ? _l : 0;
    var yAxisLabelWidth = (_m = props.yAxisLabelWidth) !== null && _m !== void 0 ? _m : (props.hideYAxisText
        ? AxesAndRulesDefaults.yAxisEmptyLabelWidth
        : AxesAndRulesDefaults.yAxisLabelWidth);
    var autoComputedSectionWidth = props.initialSpacing !== undefined
        ? (parentWidth - yAxisLabelWidth) / dataLength - props.initialSpacing
        : (parentWidth - yAxisLabelWidth) / (dataLength + 0.5);
    var autoComputedBarWidth = autoComputedSectionWidth * 0.6;
    var defaultBarWidth = adjustToWidth
        ? autoComputedBarWidth
        : BarDefaults.barWidth;
    var barWidth = (_o = props.barWidth) !== null && _o !== void 0 ? _o : defaultBarWidth;
    var autoComputedSpacing = autoComputedSectionWidth * 0.4;
    var spacing = (_p = props.spacing) !== null && _p !== void 0 ? _p : (adjustToWidth ? autoComputedSpacing : BarDefaults.spacing);
    var initialSpacing = (_q = props.initialSpacing) !== null && _q !== void 0 ? _q : spacing;
    var endSpacing = (_r = props.endSpacing) !== null && _r !== void 0 ? _r : spacing;
    var horizontal = (_s = props.horizontal) !== null && _s !== void 0 ? _s : BarDefaults.horizontal;
    var rtl = (_t = props.rtl) !== null && _t !== void 0 ? _t : BarDefaults.rtl;
    var yAxisAtTop = (_u = props.yAxisAtTop) !== null && _u !== void 0 ? _u : BarDefaults.yAxisAtTop;
    var intactTopLabel = (_v = props.intactTopLabel) !== null && _v !== void 0 ? _v : BarDefaults.intactTopLabel;
    var heightFromProps = horizontal ? props.width : props.height;
    var widthFromProps = horizontal ? props.height : props.width;
    var isAnimated = (_w = props.isAnimated) !== null && _w !== void 0 ? _w : BarDefaults.isAnimated;
    var animationDuration = (_x = props.animationDuration) !== null && _x !== void 0 ? _x : BarDefaults.animationDuration;
    // const secondaryData = getSecondaryDataWithOffsetIncluded(
    //   props.secondaryData,
    //   props.secondaryYAxis
    // )
    var lineData = useMemo(function () {
        if (!props.lineData) {
            return stackData !== null && stackData !== void 0 ? stackData : data;
        }
        if (yAxisOffset) {
            return props.lineData.map(function (item) {
                var _a;
                return (__assign(__assign({}, item), { value: ((_a = item.value) !== null && _a !== void 0 ? _a : 0) - (yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0) }));
            });
        }
        return props.lineData;
    }, [yAxisOffset, props.lineData, data, stackData]);
    var lineData2 = props.lineData2;
    var lineBehindBars = (_y = props.lineBehindBars) !== null && _y !== void 0 ? _y : BarDefaults.lineBehindBars;
    defaultLineConfig.initialSpacing = initialSpacing;
    defaultLineConfig.endIndex = lineData.length - 1;
    defaultLineConfig.animationDuration = animationDuration;
    var lineConfig = props.lineConfig
        ? getLineConfigForBarChart(props.lineConfig, initialSpacing)
        : defaultLineConfig;
    var lineConfig2 = props.lineConfig2
        ? getLineConfigForBarChart(props.lineConfig2, initialSpacing)
        : defaultLineConfig;
    var noOfSections = getNoOfSections(props.noOfSections, props.maxValue, props.stepValue);
    var containerHeight = heightFromProps !== null && heightFromProps !== void 0 ? heightFromProps : (props.stepHeight
        ? props.stepHeight * noOfSections
        : AxesAndRulesDefaults.containerHeight);
    var horizSections = [{ value: '0' }];
    var stepHeight = (_z = props.stepHeight) !== null && _z !== void 0 ? _z : containerHeight / noOfSections;
    var labelWidth = (_0 = props.labelWidth) !== null && _0 !== void 0 ? _0 : AxesAndRulesDefaults.labelWidth;
    var scrollToEnd = (_1 = props.scrollToEnd) !== null && _1 !== void 0 ? _1 : BarDefaults.scrollToEnd;
    var scrollAnimation = (_2 = props.scrollAnimation) !== null && _2 !== void 0 ? _2 : BarDefaults.scrollAnimation;
    var scrollEventThrottle = (_3 = props.scrollEventThrottle) !== null && _3 !== void 0 ? _3 : BarDefaults.scrollEventThrottle;
    var labelsExtraHeight = (_4 = props.labelsExtraHeight) !== null && _4 !== void 0 ? _4 : AxesAndRulesDefaults.labelsExtraHeight;
    var secondaryMaxItem = 0;
    var secondaryMinItem = 0;
    if (lineConfig.isSecondary) {
        lineData.forEach(function (item) {
            var _a, _b, _c, _d;
            if (((_a = item.value) !== null && _a !== void 0 ? _a : 0) > secondaryMaxItem) {
                secondaryMaxItem = (_b = item.value) !== null && _b !== void 0 ? _b : 0;
            }
            if (((_c = item.value) !== null && _c !== void 0 ? _c : 0) < secondaryMinItem) {
                secondaryMinItem = (_d = item.value) !== null && _d !== void 0 ? _d : 0;
            }
        });
    }
    var totalWidth = initialSpacing + endSpacing;
    var maxItem = 0;
    var minItem = 0;
    var minPositiveItem = 0;
    var secondaryMinPositiveItem = 0;
    if (stackData) {
        stackData.forEach(function (stackItem, index) {
            var _a, _b, _c;
            var stackSumMax = stackItem.stacks.reduce(function (acc, stack) { return acc + (stack.value >= 0 ? stack.value : 0); }, 0);
            var stackSumMin = stackItem.stacks.reduce(function (acc, stack) { return acc + (stack.value < 0 ? stack.value : 0); }, 0);
            if (stackItem.isSecondary) {
                if (stackSumMax > secondaryMaxItem) {
                    secondaryMaxItem = stackSumMax;
                }
                if (stackSumMin < secondaryMinItem) {
                    secondaryMinItem = stackSumMin;
                    secondaryMinPositiveItem =
                        secondaryMinItem > 0 ? secondaryMinItem : secondaryMinPositiveItem;
                }
            }
            else {
                if (stackSumMax > maxItem) {
                    maxItem = stackSumMax;
                }
                if (stackSumMin < minItem) {
                    minItem = stackSumMin;
                    minPositiveItem = minItem > 0 ? minItem : minPositiveItem;
                }
            }
            totalWidth +=
                ((_b = (_a = stackItem.stacks[0].barWidth) !== null && _a !== void 0 ? _a : props.barWidth) !== null && _b !== void 0 ? _b : defaultBarWidth) +
                    (index === dataLength - 1 ? 0 : (_c = stackItem.spacing) !== null && _c !== void 0 ? _c : spacing);
        });
    }
    else {
        data.forEach(function (item, index) {
            var _a, _b, _c;
            if (item.isSecondary) {
                if (item.value > secondaryMaxItem) {
                    secondaryMaxItem = item.value;
                }
                if (item.value < secondaryMinItem) {
                    secondaryMinItem = item.value;
                    secondaryMinPositiveItem =
                        secondaryMinItem > 0 ? secondaryMinItem : secondaryMinPositiveItem;
                }
            }
            else {
                if (item.value > maxItem) {
                    maxItem = item.value;
                }
                if (item.value < minItem) {
                    minItem = item.value;
                    minPositiveItem = minItem > 0 ? minItem : minPositiveItem;
                }
            }
            totalWidth +=
                ((_b = (_a = item.barWidth) !== null && _a !== void 0 ? _a : props.barWidth) !== null && _b !== void 0 ? _b : defaultBarWidth) +
                    (index === dataLength - 1 ? spacing : (_c = item.spacing) !== null && _c !== void 0 ? _c : spacing);
        });
    }
    var valuesRange = maxItem - minPositiveItem; // Diff bw largest & smallest +ve values
    var showFractionalValues = (_5 = props.showFractionalValues) !== null && _5 !== void 0 ? _5 : valuesRange <= 1;
    var roundToDigits = (_6 = props.roundToDigits) !== null && _6 !== void 0 ? _6 : (showFractionalValues ? indexOfFirstNonZeroDigit(valuesRange) + 1 : 0);
    var maxAndMin = maxAndMinUtil(maxItem, minItem, roundToDigits, showFractionalValues);
    var maxValue = getMaxValue(props.maxValue, props.stepValue, noOfSections, maxAndMin.maxItem) || 10;
    var secondaryRange = secondaryMaxItem - secondaryMinPositiveItem; // Diff bw largest & smallest +ve values
    var showSecondaryFractionalValues = (_8 = (_7 = props.secondaryYAxis) === null || _7 === void 0 ? void 0 : _7.showFractionalValues) !== null && _8 !== void 0 ? _8 : secondaryRange <= 1;
    var secondaryRoundToDigits = (_10 = (_9 = props.secondaryYAxis) === null || _9 === void 0 ? void 0 : _9.roundToDigits) !== null && _10 !== void 0 ? _10 : (showSecondaryFractionalValues
        ? indexOfFirstNonZeroDigit(secondaryRange) + 1
        : 0);
    var secondaryMaxAndMin = maxAndMinUtil(secondaryMaxItem, secondaryMinItem, secondaryRoundToDigits, showSecondaryFractionalValues);
    // const secondaryMaxValue = lineConfig.isSecondary
    //   ? typeof props.secondaryYAxis !== 'boolean'
    //     ? (props.secondaryYAxis as secondaryYAxisType).maxValue ??
    //       secondaryMaxAndMin.maxItem
    //     : secondaryMaxAndMin.maxItem
    //   : maxValue
    var secondaryMaxValue = (_12 = (_11 = props.secondaryYAxis) === null || _11 === void 0 ? void 0 : _11.maxValue) !== null && _12 !== void 0 ? _12 : secondaryMaxAndMin.maxItem;
    var mostNegativeValue = getMostNegativeValue(props.mostNegativeValue, props.negativeStepValue, props.noOfSectionsBelowXAxis, maxAndMin.minItem);
    var stepValue = (_13 = props.stepValue) !== null && _13 !== void 0 ? _13 : maxValue / noOfSections;
    var effectiveNegativeStepValue = negativeStepValue !== null && negativeStepValue !== void 0 ? negativeStepValue : stepValue;
    var noOfSectionsBelowXAxis = (_14 = props.noOfSectionsBelowXAxis) !== null && _14 !== void 0 ? _14 : (effectiveNegativeStepValue
        ? Math.round(Math.ceil(-mostNegativeValue / effectiveNegativeStepValue))
        : 0);
    var showScrollIndicator = (_15 = props.showScrollIndicator) !== null && _15 !== void 0 ? _15 : BarDefaults.showScrollIndicator;
    var side = (_16 = props.side) !== null && _16 !== void 0 ? _16 : BarDefaults.side;
    var rotateLabel = (_17 = props.rotateLabel) !== null && _17 !== void 0 ? _17 : AxesAndRulesDefaults.rotateLabel;
    var opacity = (_18 = props.opacity) !== null && _18 !== void 0 ? _18 : BarDefaults.opacity;
    var isThreeD = (_19 = props.isThreeD) !== null && _19 !== void 0 ? _19 : BarDefaults.isThreeD;
    var showXAxisIndices = (_20 = props.showXAxisIndices) !== null && _20 !== void 0 ? _20 : AxesAndRulesDefaults.showXAxisIndices;
    var xAxisIndicesHeight = (_21 = props.xAxisIndicesHeight) !== null && _21 !== void 0 ? _21 : AxesAndRulesDefaults.xAxisIndicesHeight;
    var xAxisIndicesWidth = (_22 = props.xAxisIndicesWidth) !== null && _22 !== void 0 ? _22 : AxesAndRulesDefaults.xAxisIndicesWidth;
    var xAxisIndicesColor = (_23 = props.xAxisIndicesColor) !== null && _23 !== void 0 ? _23 : AxesAndRulesDefaults.xAxisIndicesColor;
    var xAxisThickness = (_24 = props.xAxisThickness) !== null && _24 !== void 0 ? _24 : AxesAndRulesDefaults.xAxisThickness;
    var xAxisTextNumberOfLines = (_25 = props.xAxisTextNumberOfLines) !== null && _25 !== void 0 ? _25 : AxesAndRulesDefaults.xAxisTextNumberOfLines;
    var xAxisLabelsVerticalShift = (_26 = props.xAxisLabelsVerticalShift) !== null && _26 !== void 0 ? _26 : AxesAndRulesDefaults.xAxisLabelsVerticalShift;
    var xAxisLabelsAtBottom = (_27 = props.xAxisLabelsAtBottom) !== null && _27 !== void 0 ? _27 : false;
    var horizontalRulesStyle = props.horizontalRulesStyle;
    var autoShiftLabels = (_28 = props.autoShiftLabels) !== null && _28 !== void 0 ? _28 : false;
    var barBorderColor = (_29 = props.barBorderColor) !== null && _29 !== void 0 ? _29 : BarDefaults.barBorderColor;
    var extendedContainerHeight = getExtendedContainerHeightWithPadding(containerHeight, 0);
    var axesAndRulesProps = getAxesAndRulesProps(props, containerHeight, stepHeight, stepValue, noOfSections, roundToDigits, negativeStepValue !== null && negativeStepValue !== void 0 ? negativeStepValue : stepValue, secondaryMaxValue, secondaryMinItem, showSecondaryFractionalValues, secondaryRoundToDigits);
    var _75 = axesAndRulesProps.secondaryYAxisConfig, secondaryStepHeight = _75.stepHeight, secondaryStepValue = _75.stepValue, secondaryNegativeStepHeight = _75.negativeStepHeight, secondaryNegativeStepValue = _75.negativeStepValue, secondaryNoOfSectionsBelowXAxis = _75.noOfSectionsBelowXAxis;
    var primary4thQuadrantHeight = noOfSectionsBelowXAxis * ((_30 = props.negativeStepHeight) !== null && _30 !== void 0 ? _30 : stepHeight);
    var secondary4thQuadrantHeight = secondaryNoOfSectionsBelowXAxis * secondaryNegativeStepHeight;
    var fourthQuadrantHeight = Math.max(primary4thQuadrantHeight, secondary4thQuadrantHeight);
    var containerHeightIncludingBelowXAxis = extendedContainerHeight + fourthQuadrantHeight;
    var _76 = __read(useState(-1), 2), pointerIndex = _76[0], setPointerIndex = _76[1];
    var _77 = __read(useState(0), 2), pointerX = _77[0], setPointerX = _77[1];
    var _78 = __read(useState(0), 2), pointerY = _78[0], setPointerY = _78[1];
    var _79 = __read(useState(), 2), pointerItem = _79[0], setPointerItem = _79[1];
    var _80 = __read(useState(0), 2), responderStartTime = _80[0], setResponderStartTime = _80[1];
    var _81 = __read(useState(false), 2), responderActive = _81[0], setResponderActive = _81[1];
    var pointerConfig = props.pointerConfig;
    var getPointerProps = (_31 = props.getPointerProps) !== null && _31 !== void 0 ? _31 : null;
    var pointerHeight = (_32 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.height) !== null && _32 !== void 0 ? _32 : defaultPointerConfig.height;
    var pointerWidth = (_33 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.width) !== null && _33 !== void 0 ? _33 : defaultPointerConfig.width;
    var pointerRadius = (_34 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.radius) !== null && _34 !== void 0 ? _34 : defaultPointerConfig.radius;
    var pointerColor = (_35 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerColor) !== null && _35 !== void 0 ? _35 : defaultPointerConfig.pointerColor;
    var pointerComponent = (_36 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerComponent) !== null && _36 !== void 0 ? _36 : defaultPointerConfig.pointerComponent;
    var showPointerStrip = (pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.showPointerStrip) === false
        ? false
        : defaultPointerConfig.showPointerStrip;
    var pointerStripHeight = (_37 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripHeight) !== null && _37 !== void 0 ? _37 : defaultPointerConfig.pointerStripHeight;
    var pointerStripWidth = (_38 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripWidth) !== null && _38 !== void 0 ? _38 : defaultPointerConfig.pointerStripWidth;
    var pointerStripColor = (_39 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripColor) !== null && _39 !== void 0 ? _39 : defaultPointerConfig.pointerStripColor;
    var pointerStripUptoDataPoint = (_40 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerStripUptoDataPoint) !== null && _40 !== void 0 ? _40 : defaultPointerConfig.pointerStripUptoDataPoint;
    var pointerLabelComponent = (_41 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelComponent) !== null && _41 !== void 0 ? _41 : defaultPointerConfig.pointerLabelComponent;
    var stripOverPointer = (_42 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.stripOverPointer) !== null && _42 !== void 0 ? _42 : defaultPointerConfig.stripOverPointer;
    var shiftPointerLabelX = (_43 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.shiftPointerLabelX) !== null && _43 !== void 0 ? _43 : defaultPointerConfig.shiftPointerLabelX;
    var shiftPointerLabelY = (_44 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.shiftPointerLabelY) !== null && _44 !== void 0 ? _44 : defaultPointerConfig.shiftPointerLabelY;
    var pointerLabelWidth = (_45 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelWidth) !== null && _45 !== void 0 ? _45 : defaultPointerConfig.pointerLabelWidth;
    var pointerLabelHeight = (_46 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerLabelHeight) !== null && _46 !== void 0 ? _46 : defaultPointerConfig.pointerLabelHeight;
    var autoAdjustPointerLabelPosition = (_47 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.autoAdjustPointerLabelPosition) !== null && _47 !== void 0 ? _47 : defaultPointerConfig.autoAdjustPointerLabelPosition;
    var pointerVanishDelay = (_48 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerVanishDelay) !== null && _48 !== void 0 ? _48 : defaultPointerConfig.pointerVanishDelay;
    var activatePointersOnLongPress = (_49 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.activatePointersOnLongPress) !== null && _49 !== void 0 ? _49 : defaultPointerConfig.activatePointersOnLongPress;
    var activatePointersDelay = (_50 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.activatePointersDelay) !== null && _50 !== void 0 ? _50 : defaultPointerConfig.activatePointersDelay;
    var initialPointerIndex = (_51 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.initialPointerIndex) !== null && _51 !== void 0 ? _51 : defaultPointerConfig.initialPointerIndex;
    var initialPointerAppearDelay = (_52 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.initialPointerAppearDelay) !== null && _52 !== void 0 ? _52 : (isAnimated
        ? animationDuration
        : defaultPointerConfig.initialPointerAppearDelay);
    var persistPointer = (_53 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.persistPointer) !== null && _53 !== void 0 ? _53 : defaultPointerConfig.persistPointer;
    var hidePointer1 = (_54 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.hidePointer1) !== null && _54 !== void 0 ? _54 : defaultPointerConfig.hidePointer1;
    var pointerEvents = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.pointerEvents;
    var stripBehindBars = (_55 = pointerConfig === null || pointerConfig === void 0 ? void 0 : pointerConfig.stripBehindBars) !== null && _55 !== void 0 ? _55 : defaultPointerConfig.stripBehindBars;
    var disableScroll = (_56 = props.disableScroll) !== null && _56 !== void 0 ? _56 : (pointerConfig
        ? activatePointersOnLongPress
            ? !!responderActive
            : true
        : false);
    var yAxisExtraHeightAtTop = props.trimYAxisAtTop
        ? 0
        : (_57 = props.yAxisExtraHeight) !== null && _57 !== void 0 ? _57 : containerHeight / 20;
    var barInnerComponent = props.barInnerComponent;
    var localYAxisOffset1 = lineConfig.isSecondary
        ? (_59 = (_58 = props.secondaryYAxis) === null || _58 === void 0 ? void 0 : _58.yAxisOffset) !== null && _59 !== void 0 ? _59 : 0
        : yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0;
    var localYAxisOffset2 = lineConfig2.isSecondary
        ? (_61 = (_60 = props.secondaryYAxis) === null || _60 === void 0 ? void 0 : _60.yAxisOffset) !== null && _61 !== void 0 ? _61 : 0
        : yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0;
    useEffect(function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2;
        if (showLine) {
            var pp = '';
            var pp2 = '';
            var firstBarWidth = (_c = (_b = (_a = (stackData !== null && stackData !== void 0 ? stackData : data)) === null || _a === void 0 ? void 0 : _a[0].barWidth) !== null && _b !== void 0 ? _b : props.barWidth) !== null && _c !== void 0 ? _c : defaultBarWidth;
            if (!lineConfig.curved) {
                for (var i = 0; i < lineData.length; i++) {
                    if (i < ((_d = lineConfig.startIndex) !== null && _d !== void 0 ? _d : 0) ||
                        i > ((_e = lineConfig.endIndex) !== null && _e !== void 0 ? _e : 0)) {
                        continue;
                    }
                    var currentBarWidth = (_j = (_h = (_g = (_f = (stackData !== null && stackData !== void 0 ? stackData : data)) === null || _f === void 0 ? void 0 : _f[i]) === null || _g === void 0 ? void 0 : _g.barWidth) !== null && _h !== void 0 ? _h : props.barWidth) !== null && _j !== void 0 ? _j : defaultBarWidth;
                    var currentValue = props.lineData
                        ? props.lineData[i].value
                        : stackData
                            ? stackData[i].stacks.reduce(function (total, item) { return total + item.value; }, 0)
                            : data[i].value;
                    pp +=
                        'L' +
                            getXForLineInBar(i, firstBarWidth, currentBarWidth, yAxisLabelWidth, lineConfig, spacing) +
                            ' ' +
                            getYForLineInBar(currentValue, lineConfig.shiftY, containerHeight, lineConfig.isSecondary ? secondaryMaxValue : maxValue, localYAxisOffset1) +
                            ' ';
                }
                setPoints(pp.replace('L', 'M'));
                if (lineData.length > 1 && lineConfig.showArrow) {
                    var ppArray = pp.trim().split(' ');
                    var arrowTipY = parseInt(ppArray[ppArray.length - 1]);
                    var arrowTipX = parseInt(ppArray[ppArray.length - 2].replace('L', ''));
                    var y1 = parseInt(ppArray[ppArray.length - 3]);
                    var x1 = parseInt(ppArray[ppArray.length - 4].replace('L', ''));
                    var arrowPoints_1 = getArrowPoints(arrowTipX, arrowTipY, x1, y1, (_k = lineConfig.arrowConfig) === null || _k === void 0 ? void 0 : _k.length, (_l = lineConfig.arrowConfig) === null || _l === void 0 ? void 0 : _l.width, (_m = lineConfig.arrowConfig) === null || _m === void 0 ? void 0 : _m.showArrowBase);
                    setArrowPoints(arrowPoints_1);
                }
            }
            else {
                var p1Array = [];
                for (var i = 0; i < lineData.length; i++) {
                    if (i < ((_o = lineConfig.startIndex) !== null && _o !== void 0 ? _o : 0) ||
                        i > ((_p = lineConfig.endIndex) !== null && _p !== void 0 ? _p : 0)) {
                        continue;
                    }
                    var currentBarWidth = (_s = (_r = (_q = data === null || data === void 0 ? void 0 : data[i]) === null || _q === void 0 ? void 0 : _q.barWidth) !== null && _r !== void 0 ? _r : props.barWidth) !== null && _s !== void 0 ? _s : defaultBarWidth;
                    var currentValue = props.lineData
                        ? props.lineData[i].value
                        : stackData
                            ? stackData[i].stacks.reduce(function (total, item) { return total + item.value; }, 0)
                            : data[i].value;
                    p1Array.push([
                        getXForLineInBar(i, firstBarWidth, currentBarWidth, yAxisLabelWidth, lineConfig, spacing),
                        getYForLineInBar(currentValue, lineConfig.shiftY, containerHeight, lineConfig.isSecondary ? secondaryMaxValue : maxValue, localYAxisOffset1)
                    ]);
                    var xx = svgPath(p1Array, lineConfig.curveType, lineConfig.curvature);
                    setPoints(xx);
                }
            }
            if (lineData2 === null || lineData2 === void 0 ? void 0 : lineData2.length) {
                if (!(lineConfig2 === null || lineConfig2 === void 0 ? void 0 : lineConfig2.curved)) {
                    for (var i = 0; i < lineData2.length; i++) {
                        if (i < ((_t = lineConfig2.startIndex) !== null && _t !== void 0 ? _t : 0) ||
                            i > ((_u = lineConfig2.endIndex) !== null && _u !== void 0 ? _u : 0)) {
                            continue;
                        }
                        var currentBarWidth = (_x = (_w = (_v = data === null || data === void 0 ? void 0 : data[i]) === null || _v === void 0 ? void 0 : _v.barWidth) !== null && _w !== void 0 ? _w : props.barWidth) !== null && _x !== void 0 ? _x : defaultBarWidth;
                        var currentValue = lineData2[i].value;
                        pp2 +=
                            'L' +
                                getXForLineInBar(i, firstBarWidth, currentBarWidth, yAxisLabelWidth, lineConfig2, spacing) +
                                ' ' +
                                getYForLineInBar(currentValue, lineConfig2.shiftY, containerHeight, lineConfig2.isSecondary ? secondaryMaxValue : maxValue, localYAxisOffset2) +
                                ' ';
                    }
                    setPoints2(pp2.replace('L', 'M'));
                }
                else {
                    var p2Array = [];
                    for (var i = 0; i < lineData2.length; i++) {
                        if (i < ((_y = lineConfig2.startIndex) !== null && _y !== void 0 ? _y : 0) ||
                            i > ((_z = lineConfig2.endIndex) !== null && _z !== void 0 ? _z : 0)) {
                            continue;
                        }
                        var currentBarWidth = (_2 = (_1 = (_0 = data === null || data === void 0 ? void 0 : data[i]) === null || _0 === void 0 ? void 0 : _0.barWidth) !== null && _1 !== void 0 ? _1 : props.barWidth) !== null && _2 !== void 0 ? _2 : defaultBarWidth;
                        var currentValue = lineData2[i].value;
                        p2Array.push([
                            getXForLineInBar(i, firstBarWidth, currentBarWidth, yAxisLabelWidth, lineConfig2, spacing),
                            getYForLineInBar(currentValue, lineConfig2.shiftY, containerHeight, lineConfig2.isSecondary ? secondaryMaxValue : maxValue, localYAxisOffset2)
                        ]);
                        var xx = svgPath(p2Array, lineConfig2.curveType, lineConfig2.curvature);
                        setPoints2(xx);
                    }
                }
            }
        }
    }, [
        animationDuration,
        containerHeight,
        data,
        lineData,
        initialSpacing,
        lineConfig.initialSpacing,
        lineConfig.curved,
        lineConfig.dataPointsWidth,
        lineConfig.shiftY,
        lineConfig.isAnimated,
        lineConfig.delay,
        lineConfig.startIndex,
        lineConfig.endIndex,
        maxValue,
        props.barWidth,
        showLine,
        spacing,
        yAxisLabelWidth,
        lineConfig.showArrow,
        (_62 = lineConfig.arrowConfig) === null || _62 === void 0 ? void 0 : _62.length,
        (_63 = lineConfig.arrowConfig) === null || _63 === void 0 ? void 0 : _63.width,
        (_64 = lineConfig.arrowConfig) === null || _64 === void 0 ? void 0 : _64.showArrowBase
    ]);
    useEffect(function () {
        var _a, _b;
        if (initialPointerIndex !== -1) {
            var item_1 = (_a = (stackData !== null && stackData !== void 0 ? stackData : data)) === null || _a === void 0 ? void 0 : _a[initialPointerIndex];
            var stackItem = stackData === null || stackData === void 0 ? void 0 : stackData[initialPointerIndex];
            var stackSum = (_b = stackItem === null || stackItem === void 0 ? void 0 : stackItem.stacks) === null || _b === void 0 ? void 0 : _b.reduce(function (acc, stack) { var _a; return acc + ((_a = stack.value) !== null && _a !== void 0 ? _a : 0); }, 0);
            var x_1 = initialSpacing +
                (spacing + barWidth) * initialPointerIndex -
                (pointerRadius !== null && pointerRadius !== void 0 ? pointerRadius : pointerWidth / 2) +
                barWidth / 2;
            var y_1 = containerHeight -
                ((stackSum !== null && stackSum !== void 0 ? stackSum : data[initialPointerIndex].value) * containerHeight) /
                    maxValue -
                (pointerRadius !== null && pointerRadius !== void 0 ? pointerRadius : pointerHeight / 2) +
                10;
            if (initialPointerAppearDelay) {
                setTimeout(function () {
                    setPointerConfig(initialPointerIndex, item_1, x_1, y_1);
                }, initialPointerAppearDelay);
            }
            else {
                setPointerConfig(initialPointerIndex, item_1, x_1, y_1);
            }
        }
    }, []);
    var setPointerConfig = function (initialPointerIndex, item, x, y) {
        setPointerIndex(initialPointerIndex);
        setPointerItem(item);
        setPointerX(x);
        setPointerY(y);
    };
    var animatedHeight = heightValue === null || heightValue === void 0 ? void 0 : heightValue.interpolate({
        inputRange: [0, 1],
        outputRange: ['0%', '100%']
    });
    var appearingOpacity = opacityValue === null || opacityValue === void 0 ? void 0 : opacityValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, 1]
    });
    var animatedWidth = widthValue === null || widthValue === void 0 ? void 0 : widthValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, initialSpacing + totalWidth]
    });
    var getPropsCommonForBarAndStack = function (item, index) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z;
        return {
            item: item,
            index: index,
            containerHeight: containerHeight,
            containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
            maxValue: maxValue,
            spacing: (_a = item.spacing) !== null && _a !== void 0 ? _a : spacing,
            propSpacing: spacing,
            xAxisThickness: xAxisThickness,
            barWidth: (_b = props.barWidth) !== null && _b !== void 0 ? _b : defaultBarWidth,
            opacity: opacity,
            disablePress: (_c = item.disablePress) !== null && _c !== void 0 ? _c : props.disablePress,
            rotateLabel: rotateLabel,
            showXAxisIndices: showXAxisIndices,
            xAxisIndicesHeight: xAxisIndicesHeight,
            xAxisIndicesWidth: xAxisIndicesWidth,
            xAxisIndicesColor: xAxisIndicesColor,
            labelsDistanceFromXaxis: (_e = (_d = item.labelsDistanceFromXaxis) !== null && _d !== void 0 ? _d : labelsDistanceFromXaxis) !== null && _e !== void 0 ? _e : (xAxisLabelsAtBottom ? fourthQuadrantHeight : 0),
            horizontal: horizontal,
            rtl: rtl,
            intactTopLabel: intactTopLabel,
            showValuesAsTopLabel: props.showValuesAsTopLabel,
            topLabelContainerStyle: props.topLabelContainerStyle,
            topLabelTextStyle: props.topLabelTextStyle,
            barBorderWidth: props.barBorderWidth,
            barBorderColor: barBorderColor,
            barBorderRadius: props.barBorderRadius,
            barBorderTopLeftRadius: props.barBorderTopLeftRadius,
            barBorderTopRightRadius: props.barBorderTopRightRadius,
            barBorderBottomLeftRadius: props.barBorderBottomLeftRadius,
            barBorderBottomRightRadius: props.barBorderBottomRightRadius,
            barInnerComponent: barInnerComponent,
            color: props.color,
            showGradient: props.showGradient,
            gradientColor: props.gradientColor,
            barBackgroundPattern: props.barBackgroundPattern,
            patternId: props.patternId,
            onPress: props.onPress,
            onLongPress: props.onLongPress,
            onPressOut: props.onPressOut,
            onContextMenu: props.onContextMenu,
            onMouseEnter: props.onMouseEnter,
            onMouseLeave: props.onMouseLeave,
            focusBarOnPress: props.focusBarOnPress,
            focusedBarConfig: props.focusedBarConfig,
            xAxisTextNumberOfLines: xAxisTextNumberOfLines,
            xAxisLabelsHeight: props.xAxisLabelsHeight,
            xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
            xAxisLabelsAtBottom: xAxisLabelsAtBottom,
            renderTooltip: props.renderTooltip,
            renderTooltipConditions: (_f = props.renderTooltipConditions) !== null && _f !== void 0 ? _f : BarDefaults.renderTooltipConditions,
            leftShiftForTooltip: (_g = props.leftShiftForTooltip) !== null && _g !== void 0 ? _g : 0,
            autoCenterTooltip: autoCenterTooltip,
            initialSpacing: initialSpacing,
            selectedIndex: selectedIndex,
            setSelectedIndex: setSelectedIndex,
            activeOpacity: (_h = props.activeOpacity) !== null && _h !== void 0 ? _h : 0.2,
            noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
            leftShiftForLastIndexTooltip: (_j = props.leftShiftForLastIndexTooltip) !== null && _j !== void 0 ? _j : 0,
            label: (_k = item.label) !== null && _k !== void 0 ? _k : (((_l = props.xAxisLabelTexts) === null || _l === void 0 ? void 0 : _l[index]) ? props.xAxisLabelTexts[index] : ''),
            secondaryLabel: (_q = (_m = item.secondaryLabel) !== null && _m !== void 0 ? _m : (_p = (_o = props.secondaryXAxis) === null || _o === void 0 ? void 0 : _o.labelTexts) === null || _p === void 0 ? void 0 : _p[index]) !== null && _q !== void 0 ? _q : '',
            labelTextStyle: (_r = item.labelTextStyle) !== null && _r !== void 0 ? _r : props.xAxisLabelTextStyle,
            secondaryLabelTextStyle: (_v = (_u = (_s = item.secondaryLabelTextStyle) !== null && _s !== void 0 ? _s : (_t = props.secondaryXAxis) === null || _t === void 0 ? void 0 : _t.labelsTextStyle) !== null && _u !== void 0 ? _u : item.labelTextStyle) !== null && _v !== void 0 ? _v : props.xAxisLabelTextStyle,
            pointerConfig: pointerConfig,
            yAxisExtraHeightAtTop: yAxisExtraHeightAtTop,
            yAxisOffset: yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0,
            focusedBarIndex: focusedBarIndex,
            stepHeight: stepHeight,
            stepValue: stepValue,
            negativeStepHeight: (_w = props.negativeStepHeight) !== null && _w !== void 0 ? _w : stepHeight,
            negativeStepValue: (_x = props.negativeStepValue) !== null && _x !== void 0 ? _x : stepValue,
            secondaryXAxis: props.secondaryXAxis,
            secondaryYAxis: props.secondaryYAxis,
            secondaryStepHeight: secondaryStepHeight,
            secondaryStepValue: secondaryStepValue,
            secondaryNegativeStepHeight: secondaryNegativeStepHeight,
            secondaryNegativeStepValue: secondaryNegativeStepValue,
            secondaryNoOfSectionsBelowXAxis: secondaryNoOfSectionsBelowXAxis,
            barMarginBottom: (_z = (_y = item.barMarginBottom) !== null && _y !== void 0 ? _y : props.barMarginBottom) !== null && _z !== void 0 ? _z : 0,
            highlightEnabled: highlightEnabled,
            highlightedBarIndex: highlightedBarIndex,
            lowlightOpacity: lowlightOpacity,
            stackHighlightEnabled: stackHighlightEnabled
        };
    };
    var barAndLineChartsWrapperProps = {
        chartType: chartTypes.BAR,
        containerHeight: containerHeight,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: stepHeight,
        negativeStepHeight: (_65 = props.negativeStepHeight) !== null && _65 !== void 0 ? _65 : stepHeight,
        labelsExtraHeight: labelsExtraHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        rtl: rtl,
        shiftX: (_66 = props.shiftX) !== null && _66 !== void 0 ? _66 : 0,
        shiftY: (_67 = props.shiftY) !== null && _67 !== void 0 ? _67 : 0,
        yAxisAtTop: yAxisAtTop,
        initialSpacing: initialSpacing,
        data: data,
        stackData: stackData,
        // secondaryData,
        barWidth: (_68 = props.barWidth) !== null && _68 !== void 0 ? _68 : defaultBarWidth,
        xAxisThickness: xAxisThickness,
        totalWidth: totalWidth,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollToIndex: props.scrollToIndex,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        indicatorColor: props.indicatorColor,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        spacing: spacing,
        showLine: showLine,
        lineConfig: lineConfig,
        lineConfig2: lineConfig2,
        maxValue: maxValue,
        lineData: lineData,
        lineData2: lineData2,
        animatedWidth: animatedWidth,
        lineBehindBars: lineBehindBars,
        points: points,
        points2: points2,
        arrowPoints: arrowPoints,
        // horizSectionProps-
        width: widthFromProps,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        sectionColors: props.sectionColors,
        showFractionalValues: showFractionalValues,
        axesAndRulesProps: axesAndRulesProps,
        yAxisLabelTexts: props.yAxisLabelTexts,
        yAxisOffset: yAxisOffset,
        rotateYAxisTexts: props.rotateYAxisTexts,
        hideAxesAndRules: props.hideAxesAndRules,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        // These are Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        pointerConfig: pointerConfig,
        getPointerProps: getPointerProps,
        pointerIndex: pointerIndex,
        pointerX: pointerX,
        pointerY: pointerY,
        onEndReached: props.onEndReached,
        onStartReached: props.onStartReached,
        endReachedOffset: (_69 = props.endReachedOffset) !== null && _69 !== void 0 ? _69 : BarDefaults.endReachedOffset,
        onMomentumScrollEnd: props.onMomentumScrollEnd,
        customBackground: props.customBackground,
        highlightEnabled: highlightEnabled,
        lowlightOpacity: lowlightOpacity,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        onScrollEndDrag: props.onScrollEndDrag,
        floatingYAxisLabels: floatingYAxisLabels,
        allowFontScaling: allowFontScaling
    };
    return {
        lineConfig: lineConfig,
        hidePointer1: hidePointer1,
        pointerItem: pointerItem,
        pointerY: pointerY,
        pointerConfig: pointerConfig,
        pointerColor: pointerColor,
        pointerX: pointerX,
        pointerComponent: pointerComponent,
        pointerHeight: pointerHeight,
        pointerRadius: pointerRadius,
        pointerWidth: pointerWidth,
        autoAdjustPointerLabelPosition: autoAdjustPointerLabelPosition,
        pointerLabelWidth: pointerLabelWidth,
        activatePointersOnLongPress: activatePointersOnLongPress,
        yAxisLabelWidth: yAxisLabelWidth,
        shiftPointerLabelX: shiftPointerLabelX,
        pointerLabelHeight: pointerLabelHeight,
        pointerStripUptoDataPoint: pointerStripUptoDataPoint,
        pointerStripHeight: pointerStripHeight,
        shiftPointerLabelY: shiftPointerLabelY,
        showPointerStrip: showPointerStrip,
        pointerStripWidth: pointerStripWidth,
        containerHeight: containerHeight,
        xAxisThickness: xAxisThickness,
        pointerStripColor: pointerStripColor,
        pointerEvents: pointerEvents,
        setResponderStartTime: setResponderStartTime,
        setPointerY: setPointerY,
        setPointerItem: setPointerItem,
        initialSpacing: initialSpacing,
        spacing: spacing,
        data: data,
        barWidth: barWidth,
        setPointerX: setPointerX,
        setPointerIndex: setPointerIndex,
        maxValue: maxValue,
        maxItem: maxItem,
        responderStartTime: responderStartTime,
        responderActive: responderActive,
        setResponderActive: setResponderActive,
        activatePointersDelay: activatePointersDelay,
        persistPointer: persistPointer,
        pointerVanishDelay: pointerVanishDelay,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        extendedContainerHeight: extendedContainerHeight,
        totalWidth: totalWidth,
        stripBehindBars: stripBehindBars,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: stepHeight,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        labelsExtraHeight: labelsExtraHeight,
        stripOverPointer: stripOverPointer,
        pointerLabelComponent: pointerLabelComponent,
        opacity: opacity,
        rotateLabel: rotateLabel,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        autoShiftLabelsForNegativeStacks: autoShiftLabelsForNegativeStacks,
        horizontal: horizontal,
        rtl: rtl,
        intactTopLabel: intactTopLabel,
        barBorderColor: barBorderColor,
        barInnerComponent: barInnerComponent,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        isAnimated: isAnimated,
        animationDuration: animationDuration,
        side: side,
        labelWidth: labelWidth,
        isThreeD: isThreeD,
        animatedHeight: animatedHeight,
        appearingOpacity: appearingOpacity,
        autoShiftLabels: autoShiftLabels,
        yAxisAtTop: yAxisAtTop,
        // secondaryData,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        showLine: showLine,
        lineConfig2: lineConfig2,
        lineData: lineData,
        lineData2: lineData2,
        animatedWidth: animatedWidth,
        lineBehindBars: lineBehindBars,
        points: points,
        setPoints: setPoints,
        points2: points2,
        setPoints2: setPoints2,
        arrowPoints: arrowPoints,
        setArrowPoints: setArrowPoints,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        showFractionalValues: showFractionalValues,
        widthFromProps: widthFromProps,
        stepValue: stepValue,
        secondaryMaxValue: secondaryMaxValue,
        getPointerProps: getPointerProps,
        pointerIndex: pointerIndex,
        getPropsCommonForBarAndStack: getPropsCommonForBarAndStack,
        barAndLineChartsWrapperProps: barAndLineChartsWrapperProps,
        yAxisExtraHeightAtTop: yAxisExtraHeightAtTop,
        selectedStackIndex: selectedStackIndex,
        setSelectedStackIndex: setSelectedStackIndex,
        allowFontScaling: allowFontScaling
    };
};
